{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "`geneanot` provides gene and transcript annotation of vertebrates species based on [Ensembl](https://www.ensembl.org/index.html).\n",
    "\n",
    "## Species\n",
    "`geneanot` supports all vertebrates species. The default species is Homo sapiens.\n",
    "\n",
    "## Requirements\n",
    "`geneanot` requires an Ensembl annotation GFF3 file. `geneanot` provides functions to download the Ensembl annotation file to a local folder, which will be used by the package. If a local annotation file exists, `geneanot` can also check if there is a new version in Ensembl and download it.\n",
    "\n",
    "The user must designate a folder that holds the Ensembl annotation file (or that will be used by `geneanot` to download the Ensembl annotation file to). This is defined below by `Annotation_folder`.\n",
    "\n",
    "## Chromosome Data\n",
    "Some of the methods require access to the chromosome data. The two supported **access modes** are:\n",
    "1. `local` - using a local chromosome Fasta file provided by the user\n",
    "2. `remote` - using the remote Ensembl chromosome files via the Ensembl REST API\n",
    "\n",
    "The `local` access mode requires the user to provide the corresponding chromosome Fasta file (this can be done at instantiation, and/or during the lifetime of the annotation object). The Fasta file **must** contain equal number of bps per row for all sequence rows (except possibly the last row), since this is assumed by `geneanot` to support fast random access. The chromosome Fasta files can be downloaded from Ensembl. For example, for Homo spaiens, these are located in\n",
    "`https://ftp.ensembl.org/pub/current_fasta/homo_sapiens/dna/` (the required files are `Homo_sapiens.GRCh38.dna_sm.chromosome.<chromosome number>.fa.gz`, e.g. `Homo_sapiens.GRCh38.dna_sm.chromosome.1.fa.gz` for chromosome 1). The Ensembl chromosome Fasta files contain equal number of bps per row for all sequence rows.\n",
    "Since in `local` mode `geneanot` uses a local chromosome Fasta file, the sequence retrival is fast.\n",
    "\n",
    "In the `remote` access mode, `geneanot` uses the Ensembl REST API to extract sequences from the corresponding chromosome. No local chromosome Fasta file is required, however this implies a slower sequence retrival, and requires a network connection.\n",
    "\n",
    "The access mode is determined by `chrm_fasta_file` below. Setting it to `None` implies `remote` mode, otherwise, holding a valid chromosome Fasta file name implies `local` mode. Note that this may be set upon annotation class instantiation, and/or during the lifetime of the annotation object.\n",
    "\n",
    "\n",
    "The following sections provide useful usage examples.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import geneanot as u\n",
    "\n",
    "# folder that holds (or is designated to hold) the Ensembl annotation file.\n",
    "Annotation_folder: Path = Path('./../AnnotationDB')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Human Gene Annotation\n",
    "\n",
    "We start with examples of annotating Homo sapiens genes and transcripts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Informative\n",
    "This following is not required for annotation. It just provides information about the annotation file in the `Annotation_folder` folder and the latest release in Ensembl."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1 annotation files in ../AnnotationDB. releases=[113], latest release = 113.\n",
      "Ensembl latest release is 113.\n"
     ]
    }
   ],
   "source": [
    "# Check for Ensembl annotation files in the Annotation_folder\n",
    "if (local_files := u.get_annotation_files_info_in_folder(Annotation_folder)):\n",
    "    releases = list(local_files.keys())\n",
    "    print(f\"Found {len(local_files)} annotation files in {Annotation_folder}. {releases=}, latest release = {max(releases)}.\")\n",
    "else:\n",
    "    print(f\"No annotation file found in {Annotation_folder=}.\")\n",
    "\n",
    "# check the latest Ensembl release\n",
    "ensembl_release = u.get_ensembl_release()\n",
    "print(f\"Ensembl latest release is {ensembl_release}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialize the Annotation Folder\n",
    "This is required if:\n",
    "1. `Annotation_folder` does not contain an annotation file, as it will be downloaded from the Ensembl site. Or,\n",
    "1. we want to update the annotation file in `Annotation_folder` to the lastest Ensembl release\n",
    "\n",
    "Otherwise, this can be ignored, and `annotation_full_file` must be set by the user, and must point to a valid annotation file name. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Local annotation releases found: 113. Latest is 113 (Homo_sapiens.GRCh38.113.gff3.gz).\n",
      "Ensembl latest release is 113.\n",
      "Ensembl and local releases match.\n",
      "Annotation file that will be used: ../AnnotationDB/Homo_sapiens.GRCh38.113.gff3.gz.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "enable_download: bool = True  # enables/disables downloading the annotation file from Ensembl \n",
    "                              # only when local release < ensembl latest release.\n",
    "                              # However, if Annotation_folder does not contain an anotation file, \n",
    "                              # the download is done regardless of this setting.\n",
    "# -----------------------------------------------------------------------------------------------\n",
    "download_done, ensembl_file, local_file = u.update_local_release_to_latest(Annotation_folder, enable_download=enable_download)\n",
    "annotation_full_file = Annotation_folder / (ensembl_file if download_done else local_file)\n",
    "print(f\"Annotation file that will be used: {annotation_full_file}.\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Instantiate an Annotation Object\n",
    "This requires the annotation file defined in `annotation_full_file` (see above).\n",
    "\n",
    "The instantiation is slow, as `geneanot` starts by preparing several data structures from the annotation file. Below we give an example of a faster way to instantiate annotation objects, which is more relevant when annotating multiple genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading ../AnnotationDB/Homo_sapiens.GRCh38.113.gff3.gz to a dataframe ... Done.\n"
     ]
    }
   ],
   "source": [
    "gene: str = 'EGFR'  # set a gene name or a gene ID\n",
    "verbose: bool = True  # set to True to get informative prints\n",
    "# -----------------------------------------------------------\n",
    "# instantiate in remote access mode\n",
    "gA = u.Gene_cls(gene, annotation_full_file, verbose=verbose)\n",
    "\n",
    "# Or - instantiate in local access mode\n",
    "# chrm_fasta_file: str = './../Chromosome/Homo_sapiens.GRCh38.dna_sm.chromosome.7.fa'\n",
    "# gA = u.Gene_cls(gene, annotation_full_file, chrm_fasta_file=chrm_fasta_file, verbose=verbose)\n",
    "\n",
    "# regardless, later we can set gA.chrm_fasta_file to either None (remote access)\n",
    "# or a valid chromosome Fasta file name (local access)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### General Gene Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene=EGFR, Gene name=EGFR, Gene ID=ENSG00000146648\n",
      "Source=../AnnotationDB/Homo_sapiens.GRCh38.113.gff3.gz\n",
      "Positive strand=True.\n",
      "Type=protein_coding, version=21.\n",
      "Description=epidermal growth factor receptor [Source:HGNC Symbol%3BAcc:HGNC:3236]\n",
      "Gene region=chr7:55,019,017-55,211,628.\n",
      "13 transcripts.\n",
      "\n",
      "EGFR contains 13 transcripts.\n",
      "EGFR transcript IDs:\n",
      " 1. ENST00000344576\n",
      " 2. ENST00000275493\n",
      " 3. ENST00000455089\n",
      " 4. ENST00000342916\n",
      " 5. ENST00000420316\n",
      " 6. ENST00000700144\n",
      " 7. ENST00000459688\n",
      " 8. ENST00000463948\n",
      " 9. ENST00000450046\n",
      "10. ENST00000700145\n",
      "11. ENST00000485503\n",
      "12. ENST00000700146\n",
      "13. ENST00000700147\n",
      "\n",
      "Gene start = 55,019,017, gene end = 55,211,628.\n",
      "\n",
      "gene='EGFR' encoded in chromosome 7 on the positive DNA strand.\n"
     ]
    }
   ],
   "source": [
    "gA.info()\n",
    "\n",
    "print(f\"\\n{gene} contains {len(gA)} transcripts.\")\n",
    "gA.show_all_transcript_IDs()\n",
    "\n",
    "print(f\"\\nGene start = {gA.gene_start:,}, gene end = {gA.gene_end:,}.\")\n",
    "\n",
    "print(f\"\\n{gene=} encoded in chromosome {gA.chrm} on the {'negative' if gA.rev else 'positive'} DNA strand.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EGFR\n",
      "ENSG00000146648\n",
      "epidermal growth factor receptor [Source:HGNC Symbol%3BAcc:HGNC:3236]\n",
      "protein_coding\n",
      "21\n",
      "False\n",
      "7\n",
      "55019017\n",
      "55211628\n",
      "13\n"
     ]
    }
   ],
   "source": [
    "# the various gene attributes printed above are also accessible by gA\n",
    "print(gA.gene_name, gA.gene_ID, gA.gene_desc, gA.gene_type, gA.gene_ver, gA.rev, gA.chrm, gA.gene_start, gA.gene_end, len(gA.transcripts_info), sep='\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transcript Annotation\n",
    "\n",
    "Here we annotate a transcript of `gene`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENST00000275493 information:\n",
      "Name=EGFR-201, ID=ENST00000275493.7, version=7\n",
      "Type=mRNA, biotype=protein_coding\n",
      "Start=55,019,017, end=55,211,628\n",
      "protein ID = ENSP00000275493\n",
      "ccdsid=CCDS5514.1\n",
      "ORF start offset = 261, ORF end offset = 3,893 (0-based offsets from the beginning of the RNA)\n",
      "ORF start chromosome position = 55,019,278, ORF end chromosome position = 55,205,617\n",
      "Exon regions:\n",
      "\t 1. 55,019,017 - 55,019,365, ID=ENSE00001841347, phase=-1, end phase=1\n",
      "\t 2. 55,142,286 - 55,142,437, ID=ENSE00003541288, phase=1, end phase=0\n",
      "\t 3. 55,143,305 - 55,143,488, ID=ENSE00001704157, phase=0, end phase=1\n",
      "\t 4. 55,146,606 - 55,146,740, ID=ENSE00001798125, phase=1, end phase=1\n",
      "\t 5. 55,151,294 - 55,151,362, ID=ENSE00001683983, phase=1, end phase=1\n",
      "\t 6. 55,152,546 - 55,152,664, ID=ENSE00001652975, phase=1, end phase=0\n",
      "\t 7. 55,154,011 - 55,154,152, ID=ENSE00001623732, phase=0, end phase=1\n",
      "\t 8. 55,155,830 - 55,155,946, ID=ENSE00001751179, phase=1, end phase=1\n",
      "\t 9. 55,156,533 - 55,156,659, ID=ENSE00001084929, phase=1, end phase=2\n",
      "\t10. 55,156,759 - 55,156,832, ID=ENSE00001084931, phase=2, end phase=1\n",
      "\t11. 55,157,663 - 55,157,753, ID=ENSE00001084926, phase=1, end phase=2\n",
      "\t12. 55,160,139 - 55,160,338, ID=ENSE00001084941, phase=2, end phase=1\n",
      "\t13. 55,161,499 - 55,161,631, ID=ENSE00001084939, phase=1, end phase=2\n",
      "\t14. 55,163,733 - 55,163,823, ID=ENSE00001084927, phase=2, end phase=0\n",
      "\t15. 55,165,280 - 55,165,437, ID=ENSE00001627115, phase=0, end phase=2\n",
      "\t16. 55,171,175 - 55,171,213, ID=ENSE00001768076, phase=2, end phase=2\n",
      "\t17. 55,172,983 - 55,173,124, ID=ENSE00002684637, phase=2, end phase=0\n",
      "\t18. 55,173,921 - 55,174,043, ID=ENSE00001778519, phase=0, end phase=0\n",
      "\t19. 55,174,722 - 55,174,820, ID=ENSE00001756460, phase=0, end phase=0\n",
      "\t20. 55,181,293 - 55,181,478, ID=ENSE00001601336, phase=0, end phase=0\n",
      "\t21. 55,191,719 - 55,191,874, ID=ENSE00001681524, phase=0, end phase=0\n",
      "\t22. 55,192,766 - 55,192,841, ID=ENSE00001631695, phase=0, end phase=1\n",
      "\t23. 55,198,717 - 55,198,863, ID=ENSE00003625684, phase=1, end phase=1\n",
      "\t24. 55,200,316 - 55,200,413, ID=ENSE00001790701, phase=1, end phase=0\n",
      "\t25. 55,201,188 - 55,201,355, ID=ENSE00001801208, phase=0, end phase=0\n",
      "\t26. 55,201,735 - 55,201,782, ID=ENSE00001773562, phase=0, end phase=0\n",
      "\t27. 55,202,517 - 55,202,625, ID=ENSE00001795780, phase=0, end phase=1\n",
      "\t28. 55,205,256 - 55,211,628, ID=ENSE00001245887, phase=1, end phase=-1\n",
      "CDS regions:\n",
      "\t 1. 55,019,278 - 55,019,365, phase=0\n",
      "\t 2. 55,142,286 - 55,142,437, phase=2\n",
      "\t 3. 55,143,305 - 55,143,488, phase=0\n",
      "\t 4. 55,146,606 - 55,146,740, phase=2\n",
      "\t 5. 55,151,294 - 55,151,362, phase=2\n",
      "\t 6. 55,152,546 - 55,152,664, phase=2\n",
      "\t 7. 55,154,011 - 55,154,152, phase=0\n",
      "\t 8. 55,155,830 - 55,155,946, phase=2\n",
      "\t 9. 55,156,533 - 55,156,659, phase=2\n",
      "\t10. 55,156,759 - 55,156,832, phase=1\n",
      "\t11. 55,157,663 - 55,157,753, phase=2\n",
      "\t12. 55,160,139 - 55,160,338, phase=1\n",
      "\t13. 55,161,499 - 55,161,631, phase=2\n",
      "\t14. 55,163,733 - 55,163,823, phase=1\n",
      "\t15. 55,165,280 - 55,165,437, phase=0\n",
      "\t16. 55,171,175 - 55,171,213, phase=1\n",
      "\t17. 55,172,983 - 55,173,124, phase=1\n",
      "\t18. 55,173,921 - 55,174,043, phase=0\n",
      "\t19. 55,174,722 - 55,174,820, phase=0\n",
      "\t20. 55,181,293 - 55,181,478, phase=0\n",
      "\t21. 55,191,719 - 55,191,874, phase=0\n",
      "\t22. 55,192,766 - 55,192,841, phase=0\n",
      "\t23. 55,198,717 - 55,198,863, phase=2\n",
      "\t24. 55,200,316 - 55,200,413, phase=2\n",
      "\t25. 55,201,188 - 55,201,355, phase=0\n",
      "\t26. 55,201,735 - 55,201,782, phase=0\n",
      "\t27. 55,202,517 - 55,202,625, phase=0\n",
      "\t28. 55,205,256 - 55,205,617, phase=2\n",
      "5UTR regions:\n",
      "\t 1. 55,019,017 - 55,019,277\n",
      "3UTR regions:\n",
      "\t 1. 55,205,618 - 55,211,628\n",
      "\n",
      " EGFR-201 ENST00000275493.7 7\n",
      "\n",
      "The transcript accessible features are:\n",
      " 1. transcript_name\n",
      " 2. transcript_v_id\n",
      " 3. transcript_ccdsid\n",
      " 4. transcript_start\n",
      " 5. transcript_end\n",
      " 6. transcript_ver\n",
      " 7. transcript_type\n",
      " 8. transcript_biotype\n",
      " 9. exon_start\n",
      "10. exon_end\n",
      "11. exon_id\n",
      "12. exon_phase\n",
      "13. exon_end_phase\n",
      "14. exon_ver\n",
      "15. CDS_start\n",
      "16. CDS_end\n",
      "17. CDS_phase\n",
      "18. protein_id\n",
      "19. 5UTR_start\n",
      "20. 5UTR_end\n",
      "21. 3UTR_start\n",
      "22. 3UTR_end\n",
      "23. ORF_start_offset\n",
      "24. ORF_end_offset\n",
      "25. ORF_start_chrm_pos\n",
      "26. ORF_end_chrm_pos\n",
      "\n"
     ]
    }
   ],
   "source": [
    "transcript_id: str = 'ENST00000275493'  # transcript ID without version\n",
    "# ---------------------------------------------------------------------\n",
    "# print transcript information\n",
    "gA.transcript_info(transcript_id, verbose=True)\n",
    "\n",
    "# the transcript features are accessible via gA.transcripts_info[transcript_id], for example:\n",
    "t_info = gA.transcripts_info[transcript_id]\n",
    "print('\\n', t_info['transcript_name'], t_info['transcript_v_id'], t_info['transcript_ver'])\n",
    "\n",
    "# all available transcript features (i.e., keys in t_info)\n",
    "available_transcript_features = ''.join(f\"{i:2}. {feature}\\n\" for i, feature in zip(range(1, len(t_info)+1), t_info.keys()))\n",
    "print(f\"\\nThe transcript accessible features are:\\n{available_transcript_features}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "exon phases: \n",
    "- exon end phase is the exon phase of the next exon.\n",
    "- conversion from exon phase to CDS phase: 0-->0, 1-->2, 2-->1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>region</th>\n",
       "      <th>region_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Exon 1</td>\n",
       "      <td>(55019017, 55019365)</td>\n",
       "      <td>349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Intron 1</td>\n",
       "      <td>(55019366, 55142285)</td>\n",
       "      <td>122920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Exon 2</td>\n",
       "      <td>(55142286, 55142437)</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Intron 2</td>\n",
       "      <td>(55142438, 55143304)</td>\n",
       "      <td>867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Exon 3</td>\n",
       "      <td>(55143305, 55143488)</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Intron 3</td>\n",
       "      <td>(55143489, 55146605)</td>\n",
       "      <td>3117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Exon 4</td>\n",
       "      <td>(55146606, 55146740)</td>\n",
       "      <td>135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Intron 4</td>\n",
       "      <td>(55146741, 55151293)</td>\n",
       "      <td>4553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Exon 5</td>\n",
       "      <td>(55151294, 55151362)</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Intron 5</td>\n",
       "      <td>(55151363, 55152545)</td>\n",
       "      <td>1183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Exon 6</td>\n",
       "      <td>(55152546, 55152664)</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Intron 6</td>\n",
       "      <td>(55152665, 55154010)</td>\n",
       "      <td>1346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Exon 7</td>\n",
       "      <td>(55154011, 55154152)</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Intron 7</td>\n",
       "      <td>(55154153, 55155829)</td>\n",
       "      <td>1677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Exon 8</td>\n",
       "      <td>(55155830, 55155946)</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Intron 8</td>\n",
       "      <td>(55155947, 55156532)</td>\n",
       "      <td>586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Exon 9</td>\n",
       "      <td>(55156533, 55156659)</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Intron 9</td>\n",
       "      <td>(55156660, 55156758)</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Exon 10</td>\n",
       "      <td>(55156759, 55156832)</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Intron 10</td>\n",
       "      <td>(55156833, 55157662)</td>\n",
       "      <td>830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Exon 11</td>\n",
       "      <td>(55157663, 55157753)</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Intron 11</td>\n",
       "      <td>(55157754, 55160138)</td>\n",
       "      <td>2385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Exon 12</td>\n",
       "      <td>(55160139, 55160338)</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Intron 12</td>\n",
       "      <td>(55160339, 55161498)</td>\n",
       "      <td>1160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Exon 13</td>\n",
       "      <td>(55161499, 55161631)</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Intron 13</td>\n",
       "      <td>(55161632, 55163732)</td>\n",
       "      <td>2101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Exon 14</td>\n",
       "      <td>(55163733, 55163823)</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Intron 14</td>\n",
       "      <td>(55163824, 55165279)</td>\n",
       "      <td>1456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Exon 15</td>\n",
       "      <td>(55165280, 55165437)</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Intron 15</td>\n",
       "      <td>(55165438, 55171174)</td>\n",
       "      <td>5737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Exon 16</td>\n",
       "      <td>(55171175, 55171213)</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Intron 16</td>\n",
       "      <td>(55171214, 55172982)</td>\n",
       "      <td>1769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Exon 17</td>\n",
       "      <td>(55172983, 55173124)</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Intron 17</td>\n",
       "      <td>(55173125, 55173920)</td>\n",
       "      <td>796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Exon 18</td>\n",
       "      <td>(55173921, 55174043)</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Intron 18</td>\n",
       "      <td>(55174044, 55174721)</td>\n",
       "      <td>678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Exon 19</td>\n",
       "      <td>(55174722, 55174820)</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Intron 19</td>\n",
       "      <td>(55174821, 55181292)</td>\n",
       "      <td>6472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Exon 20</td>\n",
       "      <td>(55181293, 55181478)</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Intron 20</td>\n",
       "      <td>(55181479, 55191718)</td>\n",
       "      <td>10240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Exon 21</td>\n",
       "      <td>(55191719, 55191874)</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Intron 21</td>\n",
       "      <td>(55191875, 55192765)</td>\n",
       "      <td>891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Exon 22</td>\n",
       "      <td>(55192766, 55192841)</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Intron 22</td>\n",
       "      <td>(55192842, 55198716)</td>\n",
       "      <td>5875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Exon 23</td>\n",
       "      <td>(55198717, 55198863)</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Intron 23</td>\n",
       "      <td>(55198864, 55200315)</td>\n",
       "      <td>1452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Exon 24</td>\n",
       "      <td>(55200316, 55200413)</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Intron 24</td>\n",
       "      <td>(55200414, 55201187)</td>\n",
       "      <td>774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Exon 25</td>\n",
       "      <td>(55201188, 55201355)</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Intron 25</td>\n",
       "      <td>(55201356, 55201734)</td>\n",
       "      <td>379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Exon 26</td>\n",
       "      <td>(55201735, 55201782)</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Intron 26</td>\n",
       "      <td>(55201783, 55202516)</td>\n",
       "      <td>734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Exon 27</td>\n",
       "      <td>(55202517, 55202625)</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Intron 27</td>\n",
       "      <td>(55202626, 55205255)</td>\n",
       "      <td>2630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Exon 28</td>\n",
       "      <td>(55205256, 55211628)</td>\n",
       "      <td>6373</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         name                region  region_size\n",
       "0      Exon 1  (55019017, 55019365)          349\n",
       "1    Intron 1  (55019366, 55142285)       122920\n",
       "2      Exon 2  (55142286, 55142437)          152\n",
       "3    Intron 2  (55142438, 55143304)          867\n",
       "4      Exon 3  (55143305, 55143488)          184\n",
       "5    Intron 3  (55143489, 55146605)         3117\n",
       "6      Exon 4  (55146606, 55146740)          135\n",
       "7    Intron 4  (55146741, 55151293)         4553\n",
       "8      Exon 5  (55151294, 55151362)           69\n",
       "9    Intron 5  (55151363, 55152545)         1183\n",
       "10     Exon 6  (55152546, 55152664)          119\n",
       "11   Intron 6  (55152665, 55154010)         1346\n",
       "12     Exon 7  (55154011, 55154152)          142\n",
       "13   Intron 7  (55154153, 55155829)         1677\n",
       "14     Exon 8  (55155830, 55155946)          117\n",
       "15   Intron 8  (55155947, 55156532)          586\n",
       "16     Exon 9  (55156533, 55156659)          127\n",
       "17   Intron 9  (55156660, 55156758)           99\n",
       "18    Exon 10  (55156759, 55156832)           74\n",
       "19  Intron 10  (55156833, 55157662)          830\n",
       "20    Exon 11  (55157663, 55157753)           91\n",
       "21  Intron 11  (55157754, 55160138)         2385\n",
       "22    Exon 12  (55160139, 55160338)          200\n",
       "23  Intron 12  (55160339, 55161498)         1160\n",
       "24    Exon 13  (55161499, 55161631)          133\n",
       "25  Intron 13  (55161632, 55163732)         2101\n",
       "26    Exon 14  (55163733, 55163823)           91\n",
       "27  Intron 14  (55163824, 55165279)         1456\n",
       "28    Exon 15  (55165280, 55165437)          158\n",
       "29  Intron 15  (55165438, 55171174)         5737\n",
       "30    Exon 16  (55171175, 55171213)           39\n",
       "31  Intron 16  (55171214, 55172982)         1769\n",
       "32    Exon 17  (55172983, 55173124)          142\n",
       "33  Intron 17  (55173125, 55173920)          796\n",
       "34    Exon 18  (55173921, 55174043)          123\n",
       "35  Intron 18  (55174044, 55174721)          678\n",
       "36    Exon 19  (55174722, 55174820)           99\n",
       "37  Intron 19  (55174821, 55181292)         6472\n",
       "38    Exon 20  (55181293, 55181478)          186\n",
       "39  Intron 20  (55181479, 55191718)        10240\n",
       "40    Exon 21  (55191719, 55191874)          156\n",
       "41  Intron 21  (55191875, 55192765)          891\n",
       "42    Exon 22  (55192766, 55192841)           76\n",
       "43  Intron 22  (55192842, 55198716)         5875\n",
       "44    Exon 23  (55198717, 55198863)          147\n",
       "45  Intron 23  (55198864, 55200315)         1452\n",
       "46    Exon 24  (55200316, 55200413)           98\n",
       "47  Intron 24  (55200414, 55201187)          774\n",
       "48    Exon 25  (55201188, 55201355)          168\n",
       "49  Intron 25  (55201356, 55201734)          379\n",
       "50    Exon 26  (55201735, 55201782)           48\n",
       "51  Intron 26  (55201783, 55202516)          734\n",
       "52    Exon 27  (55202517, 55202625)          109\n",
       "53  Intron 27  (55202626, 55205255)         2630\n",
       "54    Exon 28  (55205256, 55211628)         6373"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Transcript table - exon and intron table.\n",
    "df = gA.exon_intron_map(transcript_id)\n",
    "display(df)\n",
    "\n",
    "# the transcript table can be written to excel, CSV or HTML file\n",
    "gA.exon_intron_map_to_csv(transcript_id, './../Reports/transcript_table.csv')\n",
    "gA.exon_intron_map_to_excel(transcript_id, './../Reports/transcript_table.xlsx', usr_desc={\"Description\": \"Transcript table\", \"Transcript\": transcript_id})\n",
    "gA.exon_intron_map_to_html(transcript_id, './../Reports/transcript_table.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_region</th>\n",
       "      <th>exon_size</th>\n",
       "      <th>mRNA_NT_region</th>\n",
       "      <th>ORF_NT_region</th>\n",
       "      <th>ORF_AA_region</th>\n",
       "      <th>next_exon_frame_alignment</th>\n",
       "      <th>exon_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>(55019017, 55019365)</td>\n",
       "      <td>349</td>\n",
       "      <td>[1, 349]</td>\n",
       "      <td>[1, 88]</td>\n",
       "      <td>[1, 30]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001841347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>(55142286, 55142437)</td>\n",
       "      <td>152</td>\n",
       "      <td>[350, 501]</td>\n",
       "      <td>[89, 240]</td>\n",
       "      <td>[30, 80]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00003541288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>(55143305, 55143488)</td>\n",
       "      <td>184</td>\n",
       "      <td>[502, 685]</td>\n",
       "      <td>[241, 424]</td>\n",
       "      <td>[81, 142]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001704157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>(55146606, 55146740)</td>\n",
       "      <td>135</td>\n",
       "      <td>[686, 820]</td>\n",
       "      <td>[425, 559]</td>\n",
       "      <td>[142, 187]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001798125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>(55151294, 55151362)</td>\n",
       "      <td>69</td>\n",
       "      <td>[821, 889]</td>\n",
       "      <td>[560, 628]</td>\n",
       "      <td>[187, 210]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001683983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>(55152546, 55152664)</td>\n",
       "      <td>119</td>\n",
       "      <td>[890, 1008]</td>\n",
       "      <td>[629, 747]</td>\n",
       "      <td>[210, 249]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001652975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>(55154011, 55154152)</td>\n",
       "      <td>142</td>\n",
       "      <td>[1009, 1150]</td>\n",
       "      <td>[748, 889]</td>\n",
       "      <td>[250, 297]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001623732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>(55155830, 55155946)</td>\n",
       "      <td>117</td>\n",
       "      <td>[1151, 1267]</td>\n",
       "      <td>[890, 1006]</td>\n",
       "      <td>[297, 336]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001751179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>(55156533, 55156659)</td>\n",
       "      <td>127</td>\n",
       "      <td>[1268, 1394]</td>\n",
       "      <td>[1007, 1133]</td>\n",
       "      <td>[336, 378]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001084929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>(55156759, 55156832)</td>\n",
       "      <td>74</td>\n",
       "      <td>[1395, 1468]</td>\n",
       "      <td>[1134, 1207]</td>\n",
       "      <td>[378, 403]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001084931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>(55157663, 55157753)</td>\n",
       "      <td>91</td>\n",
       "      <td>[1469, 1559]</td>\n",
       "      <td>[1208, 1298]</td>\n",
       "      <td>[403, 433]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001084926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>(55160139, 55160338)</td>\n",
       "      <td>200</td>\n",
       "      <td>[1560, 1759]</td>\n",
       "      <td>[1299, 1498]</td>\n",
       "      <td>[433, 500]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001084941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>(55161499, 55161631)</td>\n",
       "      <td>133</td>\n",
       "      <td>[1760, 1892]</td>\n",
       "      <td>[1499, 1631]</td>\n",
       "      <td>[500, 544]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001084939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>(55163733, 55163823)</td>\n",
       "      <td>91</td>\n",
       "      <td>[1893, 1983]</td>\n",
       "      <td>[1632, 1722]</td>\n",
       "      <td>[544, 574]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001084927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>(55165280, 55165437)</td>\n",
       "      <td>158</td>\n",
       "      <td>[1984, 2141]</td>\n",
       "      <td>[1723, 1880]</td>\n",
       "      <td>[575, 627]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001627115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>(55171175, 55171213)</td>\n",
       "      <td>39</td>\n",
       "      <td>[2142, 2180]</td>\n",
       "      <td>[1881, 1919]</td>\n",
       "      <td>[627, 640]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001768076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>(55172983, 55173124)</td>\n",
       "      <td>142</td>\n",
       "      <td>[2181, 2322]</td>\n",
       "      <td>[1920, 2061]</td>\n",
       "      <td>[640, 687]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00002684637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>(55173921, 55174043)</td>\n",
       "      <td>123</td>\n",
       "      <td>[2323, 2445]</td>\n",
       "      <td>[2062, 2184]</td>\n",
       "      <td>[688, 728]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001778519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>(55174722, 55174820)</td>\n",
       "      <td>99</td>\n",
       "      <td>[2446, 2544]</td>\n",
       "      <td>[2185, 2283]</td>\n",
       "      <td>[729, 761]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001756460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>(55181293, 55181478)</td>\n",
       "      <td>186</td>\n",
       "      <td>[2545, 2730]</td>\n",
       "      <td>[2284, 2469]</td>\n",
       "      <td>[762, 823]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001601336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>(55191719, 55191874)</td>\n",
       "      <td>156</td>\n",
       "      <td>[2731, 2886]</td>\n",
       "      <td>[2470, 2625]</td>\n",
       "      <td>[824, 875]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001681524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>(55192766, 55192841)</td>\n",
       "      <td>76</td>\n",
       "      <td>[2887, 2962]</td>\n",
       "      <td>[2626, 2701]</td>\n",
       "      <td>[876, 901]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001631695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>(55198717, 55198863)</td>\n",
       "      <td>147</td>\n",
       "      <td>[2963, 3109]</td>\n",
       "      <td>[2702, 2848]</td>\n",
       "      <td>[901, 950]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003625684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>(55200316, 55200413)</td>\n",
       "      <td>98</td>\n",
       "      <td>[3110, 3207]</td>\n",
       "      <td>[2849, 2946]</td>\n",
       "      <td>[950, 982]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001790701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>(55201188, 55201355)</td>\n",
       "      <td>168</td>\n",
       "      <td>[3208, 3375]</td>\n",
       "      <td>[2947, 3114]</td>\n",
       "      <td>[983, 1038]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001801208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26</td>\n",
       "      <td>(55201735, 55201782)</td>\n",
       "      <td>48</td>\n",
       "      <td>[3376, 3423]</td>\n",
       "      <td>[3115, 3162]</td>\n",
       "      <td>[1039, 1054]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001773562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>27</td>\n",
       "      <td>(55202517, 55202625)</td>\n",
       "      <td>109</td>\n",
       "      <td>[3424, 3532]</td>\n",
       "      <td>[3163, 3271]</td>\n",
       "      <td>[1055, 1091]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001795780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>28</td>\n",
       "      <td>(55205256, 55211628)</td>\n",
       "      <td>6373</td>\n",
       "      <td>[3533, 9905]</td>\n",
       "      <td>[3272, 3633]</td>\n",
       "      <td>[1091, 1210]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001245887</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    exon_number           exon_region  exon_size mRNA_NT_region ORF_NT_region  \\\n",
       "0             1  (55019017, 55019365)        349       [1, 349]       [1, 88]   \n",
       "1             2  (55142286, 55142437)        152     [350, 501]     [89, 240]   \n",
       "2             3  (55143305, 55143488)        184     [502, 685]    [241, 424]   \n",
       "3             4  (55146606, 55146740)        135     [686, 820]    [425, 559]   \n",
       "4             5  (55151294, 55151362)         69     [821, 889]    [560, 628]   \n",
       "5             6  (55152546, 55152664)        119    [890, 1008]    [629, 747]   \n",
       "6             7  (55154011, 55154152)        142   [1009, 1150]    [748, 889]   \n",
       "7             8  (55155830, 55155946)        117   [1151, 1267]   [890, 1006]   \n",
       "8             9  (55156533, 55156659)        127   [1268, 1394]  [1007, 1133]   \n",
       "9            10  (55156759, 55156832)         74   [1395, 1468]  [1134, 1207]   \n",
       "10           11  (55157663, 55157753)         91   [1469, 1559]  [1208, 1298]   \n",
       "11           12  (55160139, 55160338)        200   [1560, 1759]  [1299, 1498]   \n",
       "12           13  (55161499, 55161631)        133   [1760, 1892]  [1499, 1631]   \n",
       "13           14  (55163733, 55163823)         91   [1893, 1983]  [1632, 1722]   \n",
       "14           15  (55165280, 55165437)        158   [1984, 2141]  [1723, 1880]   \n",
       "15           16  (55171175, 55171213)         39   [2142, 2180]  [1881, 1919]   \n",
       "16           17  (55172983, 55173124)        142   [2181, 2322]  [1920, 2061]   \n",
       "17           18  (55173921, 55174043)        123   [2323, 2445]  [2062, 2184]   \n",
       "18           19  (55174722, 55174820)         99   [2446, 2544]  [2185, 2283]   \n",
       "19           20  (55181293, 55181478)        186   [2545, 2730]  [2284, 2469]   \n",
       "20           21  (55191719, 55191874)        156   [2731, 2886]  [2470, 2625]   \n",
       "21           22  (55192766, 55192841)         76   [2887, 2962]  [2626, 2701]   \n",
       "22           23  (55198717, 55198863)        147   [2963, 3109]  [2702, 2848]   \n",
       "23           24  (55200316, 55200413)         98   [3110, 3207]  [2849, 2946]   \n",
       "24           25  (55201188, 55201355)        168   [3208, 3375]  [2947, 3114]   \n",
       "25           26  (55201735, 55201782)         48   [3376, 3423]  [3115, 3162]   \n",
       "26           27  (55202517, 55202625)        109   [3424, 3532]  [3163, 3271]   \n",
       "27           28  (55205256, 55211628)       6373   [3533, 9905]  [3272, 3633]   \n",
       "\n",
       "   ORF_AA_region  next_exon_frame_alignment          exon_ID  \n",
       "0        [1, 30]                          2  ENSE00001841347  \n",
       "1       [30, 80]                          0  ENSE00003541288  \n",
       "2      [81, 142]                          2  ENSE00001704157  \n",
       "3     [142, 187]                          2  ENSE00001798125  \n",
       "4     [187, 210]                          2  ENSE00001683983  \n",
       "5     [210, 249]                          0  ENSE00001652975  \n",
       "6     [250, 297]                          2  ENSE00001623732  \n",
       "7     [297, 336]                          2  ENSE00001751179  \n",
       "8     [336, 378]                          1  ENSE00001084929  \n",
       "9     [378, 403]                          2  ENSE00001084931  \n",
       "10    [403, 433]                          1  ENSE00001084926  \n",
       "11    [433, 500]                          2  ENSE00001084941  \n",
       "12    [500, 544]                          1  ENSE00001084939  \n",
       "13    [544, 574]                          0  ENSE00001084927  \n",
       "14    [575, 627]                          1  ENSE00001627115  \n",
       "15    [627, 640]                          1  ENSE00001768076  \n",
       "16    [640, 687]                          0  ENSE00002684637  \n",
       "17    [688, 728]                          0  ENSE00001778519  \n",
       "18    [729, 761]                          0  ENSE00001756460  \n",
       "19    [762, 823]                          0  ENSE00001601336  \n",
       "20    [824, 875]                          0  ENSE00001681524  \n",
       "21    [876, 901]                          2  ENSE00001631695  \n",
       "22    [901, 950]                          2  ENSE00003625684  \n",
       "23    [950, 982]                          0  ENSE00001790701  \n",
       "24   [983, 1038]                          0  ENSE00001801208  \n",
       "25  [1039, 1054]                          0  ENSE00001773562  \n",
       "26  [1055, 1091]                          2  ENSE00001795780  \n",
       "27  [1091, 1210]                          0  ENSE00001245887  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# RNA table - maps the ORF and AAs to the mRNA sequence.\n",
    "df = gA.exon_map(transcript_id)\n",
    "display(df)\n",
    "\n",
    "# similarly to exon_intron_map() above, this table can also we written to excel, CSV or HTML file\n",
    "gA.exon_map_to_csv(transcript_id, './../Reports/mRNA_table.csv')\n",
    "gA.exon_map_to_excel(transcript_id, './../Reports/mRNA_table.xlsx', usr_desc={\"Description\": \"mRNA table\", \"Transcript\": transcript_id})\n",
    "gA.exon_map_to_html(transcript_id, './../Reports/mRNA_table.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here `mRNA_NT_region` gives the mRNA bp count, `ORF_NT_region` gives the ORF bp count, `ORF_AA_region` gives the protein AA count, and `next_exon_frame_alignment` gives the codon phase (i.e., number of bps required in the next exon to complete the last codon in the current exon)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sequences\n",
    "\n",
    "Retrieve sequences (e.g., pre-mRNA, mRNA, ORF, AA, UTRs)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Access mode is local.\n"
     ]
    }
   ],
   "source": [
    "# The access mode may be set/changed anytime:\n",
    "\n",
    "# set to use local mode (if instantiation was done with remote mode)\n",
    "gA.chrm_fasta_file = f'./../Chromosome/Homo_sapiens.GRCh38.dna_sm.chromosome.{gA.chrm}.fa'\n",
    "\n",
    "# Or, set to use remote mode (if instantiation was with local mode)\n",
    "# gA.chrm_fasta_file = None\n",
    "\n",
    "print(f\"Access mode is {gA.access_mode()}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The pre-mRNA sequence contains 192,612 bps.\n",
      "\n",
      "rna:\n",
      "AGACGTCCGGGCAGCCCCCGGCGCAGCGCGGCCGCAGCAGCCTCCGCCCCCCGCACGGTGTGAGCGCCCGACGCGGCCGAGGCGGCCGGAGTCCCGAGCTAGCCCCGGCGGCCGCCGCCGCCCAGACCGGACGACAGGCCACCTCGTCGGCGTCCGCCCGAGTCCCCGCCTCGCCGCCAACGCCACAACCACCGCGCACGGCCCCCTGACTCCGTCCAGTATTGATCGGGAGAGCCGGAGCGAGCTCTTCGGGGAGCAGCGATGCGACCCTCCGGGACGGCCGGGGCAGCGCTCCTGGCGCTGCTGGCTGCGCTCTGCCCGGCGAGTCGGGCTCTGGAGGAAAAGAAAGTTTGCCAAGGCACGAGTAACAAGCTCACGCAGTTGGGCACTTTTGAAGATCATTTTCTCAGCCTCCAGAGGATGTTCAATAACTGTGAGGTGGTCCTTGGGAATTTGGAAATTACCTATGTGCAGAGGAATTATGATCTTTCCTTCTTAAAGACCATCCAGGAGGTGGCTGGTTATGTCCTCATTGCCCTCAACACAGTGGAGCGAATTCCTTTGGAAAACCTGCAGATCATCAGAGGAAATATGTACTACGAAAATTCCTATGCCTTAGCAGTCTTATCTAACTATGATGCAAATAAAACCGGACTGAAGGAGCTGCCCATGAGAAATTTACAGGAAATCCTGCATGGCGCCGTGCGGTTCAGCAACAACCCTGCCCTGTGCAACGTGGAGAGCATCCAGTGGCGGGACATAGTCAGCAGTGACTTTCTCAGCAACATGTCGATGGACTTCCAGAACCACCTGGGCAGCTGCCAAAAGTGTGATCCAAGCTGTCCCAATGGGAGCTGCTGGGGTGCAGGAGAGGAGAACTGCCAGAAACTGACCAAAATCATCTGTGCCCAGCAGTGCTCCGGGCGCTGCCGTGGCAAGTCCCCCAGTGACTGCTGCCACAACCAGTGTGCTGCAGGCTGCACAGGCCCCCGGGAGAGCGACTGCCTGGTCTGCCGCAAATTCCGAGACGAAGCCACGTGCAAGGACACCTGCCCCCCACTCATGCTCTACAACCCCACCACGTACCAGATGGATGTGAACCCCGAGGGCAAATACAGCTTTGGTGCCACCTGCGTGAAGAAGTGTCCCCGTAATTATGTGGTGACAGATCACGGCTCGTGCGTCCGAGCCTGTGGGGCCGACAGCTATGAGATGGAGGAAGACGGCGTCCGCAAGTGTAAGAAGTGCGAAGGGCCTTGCCGCAAAGTGTGTAACGGAATAGGTATTGGTGAATTTAAAGACTCACTCTCCATAAATGCTACGAATATTAAACACTTCAAAAACTGCACCTCCATCAGTGGCGATCTCCACATCCTGCCGGTGGCATTTAGGGGTGACTCCTTCACACATACTCCTCCTCTGGATCCACAGGAACTGGATATTCTGAAAACCGTAAAGGAAATCACAGGGTTTTTGCTGATTCAGGCTTGGCCTGAAAACAGGACGGACCTCCATGCCTTTGAGAACCTAGAAATCATACGCGGCAGGACCAAGCAACATGGTCAGTTTTCTCTTGCAGTCGTCAGCCTGAACATAACATCCTTGGGATTACGCTCCCTCAAGGAGATAAGTGATGGAGATGTGATAATTTCAGGAAACAAAAATTTGTGCTATGCAAATACAATAAACTGGAAAAAACTGTTTGGGACCTCCGGTCAGAAAACCAAAATTATAAGCAACAGAGGTGAAAACAGCTGCAAGGCCACAGGCCAGGTCTGCCATGCCTTGTGCTCCCCCGAGGGCTGCTGGGGCCCGGAGCCCAGGGACTGCGTCTCTTGCCGGAATGTCAGCCGAGGCAGGGAATGCGTGGACAAGTGCAACCTTCTGGAGGGTGAGCCAAGGGAGTTTGTGGAGAACTCTGAGTGCATACAGTGCCACCCAGAGTGCCTGCCTCAGGCCATGAACATCACCTGCACAGGACGGGGACCAGACAACTGTATCCAGTGTGCCCACTACATTGACGGCCCCCACTGCGTCAAGACCTGCCCGGCAGGAGTCATGGGAGAAAACAACACCCTGGTCTGGAAGTACGCAGACGCCGGCCATGTGTGCCACCTGTGCCATCCAAACTGCACCTACGGATGCACTGGGCCAGGTCTTGAAGGCTGTCCAACGAATGGGCCTAAGATCCCGTCCATCGCCACTGGGATGGTGGGGGCCCTCCTCTTGCTGCTGGTGGTGGCCCTGGGGATCGGCCTCTTCATGCGAAGGCGCCACATCGTTCGGAAGCGCACGCTGCGGAGGCTGCTGCAGGAGAGGGAGCTTGTGGAGCCTCTTACACCCAGTGGAGAAGCTCCCAACCAAGCTCTCTTGAGGATCTTGAAGGAAACTGAATTCAAAAAGATCAAAGTGCTGGGCTCCGGTGCGTTCGGCACGGTGTATAAGGGACTCTGGATCCCAGAAGGTGAGAAAGTTAAAATTCCCGTCGCTATCAAGGAATTAAGAGAAGCAACATCTCCGAAAGCCAACAAGGAAATCCTCGATGAAGCCTACGTGATGGCCAGCGTGGACAACCCCCACGTGTGCCGCCTGCTGGGCATCTGCCTCACCTCCACCGTGCAGCTCATCACGCAGCTCATGCCCTTCGGCTGCCTCCTGGACTATGTCCGGGAACACAAAGACAATATTGGCTCCCAGTACCTGCTCAACTGGTGTGTGCAGATCGCAAAGGGCATGAACTACTTGGAGGACCGTCGCTTGGTGCACCGCGACCTGGCAGCCAGGAACGTACTGGTGAAAACACCGCAGCATGTCAAGATCACAGATTTTGGGCTGGCCAAACTGCTGGGTGCGGAAGAGAAAGAATACCATGCAGAAGGAGGCAAAGTGCCTATCAAGTGGATGGCATTGGAATCAATTTTACACAGAATCTATACCCACCAGAGTGATGTCTGGAGCTACGGGGTGACTGTTTGGGAGTTGATGACCTTTGGATCCAAGCCATATGACGGAATCCCTGCCAGCGAGATCTCCTCCATCCTGGAGAAAGGAGAACGCCTCCCTCAGCCACCCATATGTACCATCGATGTCTACATGATCATGGTCAAGTGCTGGATGATAGACGCAGATAGTCGCCCAAAGTTCCGTGAGTTGATCATCGAATTCTCCAAAATGGCCCGAGACCCCCAGCGCTACCTTGTCATTCAGGGGGATGAAAGAATGCATTTGCCAAGTCCTACAGACTCCAACTTCTACCGTGCCCTGATGGATGAAGAAGACATGGACGACGTGGTGGATGCCGACGAGTACCTCATCCCACAGCAGGGCTTCTTCAGCAGCCCCTCCACGTCACGGACTCCCCTCCTGAGCTCTCTGAGTGCAACCAGCAACAATTCCACCGTGGCTTGCATTGATAGAAATGGGCTGCAAAGCTGTCCCATCAAGGAAGACAGCTTCTTGCAGCGATACAGCTCAGACCCCACAGGCGCCTTGACTGAGGACAGCATAGACGACACCTTCCTCCCAGTGCCTGAATACATAAACCAGTCCGTTCCCAAAAGGCCCGCTGGCTCTGTGCAGAATCCTGTCTATCACAATCAGCCTCTGAACCCCGCGCCCAGCAGAGACCCACACTACCAGGACCCCCACAGCACTGCAGTGGGCAACCCCGAGTATCTCAACACTGTCCAGCCCACCTGTGTCAACAGCACATTCGACAGCCCTGCCCACTGGGCCCAGAAAGGCAGCCACCAAATTAGCCTGGACAACCCTGACTACCAGCAGGACTTCTTTCCCAAGGAAGCCAAGCCAAATGGCATCTTTAAGGGCTCCACAGCTGAAAATGCAGAATACCTAAGGGTCGCGCCACAAAGCAGTGAATTTATTGGAGCATGACCACGGAGGATAGTATGAGCCCTAAAAATCCAGACTCTTTCGATACCCAGGACCAAGCCACAGCAGGTCCTCCATCCCAACAGCCATGCCCGCATTAGCTCTTAGACCCACAGACTGGTTTTGCAACGTTTACACCGACTAGCCAGGAAGTACTTCCACCTCGGGCACATTTTGGGAAGTTGCATTCCTTTGTCTTCAAACTGTGAAGCATTTACAGAAACGCATCCAGCAAGAATATTGTCCCTTTGAGCAGAAATTTATCTTTCAAAGAGGTATATTTGAAAAAAAAAAAAAGTATATGTGAGGATTTTTATTGATTGGGGATCTTGGAGTTTTTCATTGTCGCTATTGATTTTTACTTCAATGGGCTCTTCCAACAAGGAAGAAGCTTGCTGGTAGCACTTGCTACCCTGAGTTCATCCAGGCCCAACTGTGAGCAAGGAGCACAAGCCACAAGTCTTCCAGAGGATGCTTGATTCCAGTGGTTCTGCTTCAAGGCTTCCACTGCAAAACACTAAAGATCCAAGAAGGCCTTCATGGCCCCAGCAGGCCGGATCGGTACTGTATCAAGTCATGGCAGGTACAGTAGGATAAGCCACTCTGTCCCTTCCTGGGCAAAGAAGAAACGGAGGGGATGGAATTCTTCCTTAGACTTACTTTTGTAAAAATGTCCCCACGGTACTTACTCCCCACTGATGGACCAGTGGTTTCCAGTCATGAGCGTTAGACTGACTTGTTTGTCTTCCATTCCATTGTTTTGAAACTCAGTATGCTGCCCCTGTCTTGCTGTCATGAAATCAGCAAGAGAGGATGACACATCAAATAATAACTCGGATTCCAGCCCACATTGGATTCATCAGCATTTGGACCAATAGCCCACAGCTGAGAATGTGGAATACCTAAGGATAGCACCGCTTTTGTTCTCGCAAAAACGTATCTCCTAATTTGAGGCTCAGATGAAATGCATCAGGTCCTTTGGGGCATAGATCAGAAGACTACAAAAATGAAGCTGCTCTGAAATCTCCTTTAGCCATCACCCCAACCCCCCAAAATTAGTTTGTGTTACTTATGGAAGATAGTTTTCTCCTTTTACTTCACTTCAAAAGCTTTTTACTCAAAGAGTATATGTTCCCTCCAGGTCAGCTGCCCCCAAACCCCCTCCTTACGCTTTGTCACACAAAAAGTGTCTCTGCCTTGAGTCATCTATTCAAGCACTTACAGCTCTGGCCACAACAGGGCATTTTACAGGTGCGAATGACAGTAGCATTATGAGTAGTGTGGAATTCAGGTAGTAAATATGAAACTAGGGTTTGAAATTGATAATGCTTTCACAACATTTGCAGATGTTTTAGAAGGAAAAAAGTTCCTTCCTAAAATAATTTCTCTACAATTGGAAGATTGGAAGATTCAGCTAGTTAGGAGCCCACCTTTTTTCCTAATCTGTGTGTGCCCTGTAACCTGACTGGTTAACAGCAGTCCTTTGTAAACAGTGTTTTAAACTCTCCTAGTCAATATCCACCCCATCCAATTTATCAAGGAAGAAATGGTTCAGAAAATATTTTCAGCCTACAGTTATGTTCAGTCACACACACATACAAAATGTTCCTTTTGCTTTTAAAGTAATTTTTGACTCCCAGATCAGTCAGAGCCCCTACAGCATTGTTAAGAAAGTATTTGATTTTTGTCTCAATGAAAATAAAACTATATTCATTTCCACTCTATTATGCTCTCAAATACCCCTAAGCATCTATACTAGCCTGGTATGGGTATGAAAGATACAAAGATAAATAAAACATAGTCCCTGATTCTAAGAAATTCACAATTTAGCAAAGGAAATGGACTCATAGATGCTAACCTTAAAACAACGTGACAAATGCCAGACAGGACCCATCAGCCAGGCACTGTGAGAGCACAGAGCAGGGAGGTTGGGTCCTGCCTGAGGAGACCTGGAAGGGAGGCCTCACAGGAGGATGACCAGGTCTCAGTCAGCGGGGAGGTGGAAAGTGCAGGTGCATCAGGGGCACCCTGACCGAGGAAACAGCTGCCAGAGGCCTCCACTGCTAAAGTCCACATAAGGCTGAGGTCAGTCACCCTAAACAACCTGCTCCCTCTAAGCCAGGGGATGAGCTTGGAGCATCCCACAAGTTCCCTAAAAGTTGCAGCCCCCAGGGGGATTTTGAGCTATCATCTCTGCACATGCTTAGTGAGAAGACTACACAACATTTCTAAGAATCTGAGATTTTATATTGTCAGTTAACCACTTTCATTATTCATTCACCTCAGGACATGCAGAAATATTTCAGTCAGAACTGGGAAACAGAAGGACCTACATTCTGCTGTCACTTATGTGTCAAGAAGCAGATGATCGATGAGGCAGGTCAGTTGTAAGTGAGTCACATTGTAGCATTAAATTCTAGTATTTTTGTAGTTTGAAACAGTAACTTAATAAAAGAGCAAAAGCTATTCTAGCTTTCTTCTTCATATTTTAATTTTCCACCATAAAGTTTAGTTGCTAAATTCTATTAATTTTAAGATTGTGCTTCCCAAAATAGTTCTCACTTCATCTGTCCAGGGAGGCACAGTTCTGTCTGGTAGAAGCCGCAAAGCCCTTAGCCTCTTCACGGATCTGGCGACTGTGATGGGCAGGTCAGGAGAGGAGCTGCCCAAAGTCCCATGATTTTCACCTAACAGCCCTGATCAGTCAGTACTCAAAGCTTGGACTCCATCCCTGAAGGTCTTCCTGATTGATAGCCTGGCCTTAATACCCTACAGAAAGCCTGTCCATTGGCTGTTTCTTCCTCAGTCAGTTCCTGGAAGACCTTACCCCATGACCCCAGCTTCAGATGTGGTCTTTGGAAACAGAGGTCGAAGGAAAGTAAGGAGCTGAGAGCTCACATTCATAGGTGCCGCCAGCCTTCGTGCATCTTCTTGCATCATCTCTAAGGAGCTCCTCTAATTACACCATGCCCGTCACCCCATGAGGGATCAGAGAAGGGATGAGTCTTCTAAACTCTATATTCGCTGTGAGTCCAGGTTGTAAGGGGGAGCACTGTGGATGCATCCTATTGCACTCCAGCTGATGACACCAAAGCTTAGGTGTTTGCTGAAAGTTCTTGATGTTGTGACTTACCACCCCTGCCTCACAACTGCAGACATAAGGGGACTATGGATTGCTTAGCAGGAAAGGCACTGGTTCTCAAGGGCGGCTGCCCTTGGGAATCTTCTGGTCCCAACCAGAAAGACTGTGGCTTGATTTTCTCAGGTGCAGCCCAGCCGTAGGGCCTTTTCAGAGCACCCCCTGGTTATTGCAACATTCATCAAAGTTTCTAGAACCTCTGGCCTAAAGGAAGGGCCTGGTGGGATCTACTTGGCACTCGCTGGGGGGCCACCCCCCAGTGCCACTCTCACTAGGCCTCTGATTGCACTTGTGTAGGATGAAGCTGGTGGGTGATGGGAACTCAGCACCTCCCCTCAGGCAGAAAAGAATCATCTGTGGAGCTTCAAAAGAAGGGGCCTGGAGTCTCTGCAGACCAATTCAACCCAAATCTCGGGGGCTCTTTCATGATTCTAATGGGCAACCAGGGTTGAAACCCTTATTTCTAGGGTCTTCAGTTGTACAAGACTGTGGGTCTGTACCAGAGCCCCCGTCAGAGTAGAATAAAAGGCTGGGTAGGGTAGAGATTCCCATGTGCAGTGGAGAGAACAATCTGCAGTCACTGATAAGCCTGAGACTTGGCTCATTTCAAAAGCGTTCAATTCATCCTCACCAGCAGTTCAGCTGGAAAGGGGCAAATACCCCCACCTGAGCTTTGAAAACGCCCTGGGACCCTCTGCATTCTCTAAGTAAGTTATAGAAACCAGTCTCTTCCCTCCTTTGTGAGTGAGCTGCTATTCCACGTAGGCAACACCTGTTGAAATTGCCCTCAATGTCTACTCTGCATTTCTTTCTTGTGATAAGCACACACTTTTATTGCAACATAATGATCTGCTCACATTTCCTTGCCTGGGGGCTGTAAAACCTTACAGAACAGAAATCCTTGCCTCTTTCACCAGCCACACCTGCCATACCAGGGGTACAGCTTTGTACTATTGAAGACACAGACAGGATTTTTAAATGTAAATCTATTTTTGTAACTTTGTTGCGGGATATAGTTCTCTTTATGTAGCACTGAACTTTGTACAATATATTTTTAGAAACTCATTTTTCTACTAAAACAAACACAGTTTACTTTAGAGAGACTGCAATAGAATCAAAATTTGAAACTGAAATCTTTGTTTAAAAGGGTTAAGTTGAGGCAAGAGGAAAGCCCTTTCTCTCTCTTATAAAAAGGCACAACCTCATTGGGGAGCTAAGCTAGGTCATTGTCATGGTGAAGAAGAGAAGCATCGTTTTTATATTTAGGAAATTTTAAAAGATGATGGAAAGCACATTTAGCTTGGTCTGAGGCAGGTTCTGTTGGGGCAGTGTTAATGGAAAGGGCTCACTGTTGTTACTACTAGAAAAATCCAGTTGCATGCCATACTCTCATCATCTGCCAGTGTAACCCTGTACATGTAAGAAAAGCAATAACATAGCACTTTGTTGGTTTATATATATAATGTGACTTCAATGCAAATTTTATTTTTATATTTACAATTGATATGCATTTACCAGTATAAACTAGACATGTCTGGAGAGCCTAATAATGTTCAGCACACTTTGGTTAGTTCACCAACAGTCTTACCAAGCCTGGGCCCAGCCACCCTAGAGAAGTTATTCAGCCCTGGCTGCAGTGACATCACCTGAGGAGCTTTTAAAAGCTTGAAGCCCAGCTACACCTCAGACCGATTAAACGCAAATCTCTGGGGCTGAAACCCAAGCATTCGTAGTTTTTAAAGCTCCTGAGGTCATTCCAATGTGCGGCCAAAGTTGAGAACTACTGGCCTAGGGATTAGCCACAAGGACATGGACTTGGAGGCAAATTCTGCAGGTGTATGTGATTCTCAGGCCTAGAGAGCTAAGACACAAAGACCTCCACATCTGTCGCTGAGAGTCAAGAACCTGAACAGAGTTTCCATGAAGGTTCTCCAAGCACTAGAAGGGAGAGTGTCTAAACAATGGTTGAAAAGCAAAGGAAATATAAAACAGACACCTCTTTCCATTTCCTAAGGTTTCTCTCTTTATTAAGGGTGGACTAGTAATAAAATATAATATTCTTGCTGCTTATGCAGCTGACATTGTTGCCCTCCCTAAAGCAACCAAGTAGCCTTTATTTCCCACAGTGAAAGAAAACGCTGGCCTATCAGTTACATTACAAAAGGCAGATTTCAAGAGGATTGAGTAAGTAGTTGGATGGCTTTCATAAAAACAAGAATTCAAGAAGAGGATTCATGCTTTAAGAAACATTTGTTATACATTCCTCACAAATTATACCTGGGATAAAAACTATGTAGCAGGCAGTGTGTTTTCCTTCCATGTCTCTCTGCACTACCTGCAGTGTGTCCTCTGAGGCTGCAAGTCTGTCCTATCTGAATTCCCAGCAGAAGCACTAAGAAGCTCCACCCTATCACCTAGCAGATAAAACTATGGGGAAAACTTAAATCTGTGCATACATTTCTGGATGCATTTACTTATCTTTAAAAAAAAAGGAATCCTATGACCTGATTTGGCCACAAAAATAATCTTGCTGTACAATACAATCTCTTGGAAATTAAGAGATCCTATGGATTTGATGACTGGTATTAGAGGTGACAATGTAACCGATTAACAACAGACAGCAATAACTTCGTTTTAGAAACATTCAAGCAATAGCTTTATAGCTTCAACATATGGTACGTTTTAACCTTGAAAGTTTTGCAATGATGAAAGCAGTATTTGTACAAATGAAAAGCAGAATTCTCTTTTATATGGTTTATACTGTTGATCAGAAATGTTGATTGTGCATTGAGTATTAAAAAATTAGATGTATATTATTCATTGTTCTTTACTCCTGAGTACCTTATAATAATAATAATGTATTCTTTGTTAACAA\n",
      "9,905 bps.\n",
      "\n",
      "orf:\n",
      "ATGCGACCCTCCGGGACGGCCGGGGCAGCGCTCCTGGCGCTGCTGGCTGCGCTCTGCCCGGCGAGTCGGGCTCTGGAGGAAAAGAAAGTTTGCCAAGGCACGAGTAACAAGCTCACGCAGTTGGGCACTTTTGAAGATCATTTTCTCAGCCTCCAGAGGATGTTCAATAACTGTGAGGTGGTCCTTGGGAATTTGGAAATTACCTATGTGCAGAGGAATTATGATCTTTCCTTCTTAAAGACCATCCAGGAGGTGGCTGGTTATGTCCTCATTGCCCTCAACACAGTGGAGCGAATTCCTTTGGAAAACCTGCAGATCATCAGAGGAAATATGTACTACGAAAATTCCTATGCCTTAGCAGTCTTATCTAACTATGATGCAAATAAAACCGGACTGAAGGAGCTGCCCATGAGAAATTTACAGGAAATCCTGCATGGCGCCGTGCGGTTCAGCAACAACCCTGCCCTGTGCAACGTGGAGAGCATCCAGTGGCGGGACATAGTCAGCAGTGACTTTCTCAGCAACATGTCGATGGACTTCCAGAACCACCTGGGCAGCTGCCAAAAGTGTGATCCAAGCTGTCCCAATGGGAGCTGCTGGGGTGCAGGAGAGGAGAACTGCCAGAAACTGACCAAAATCATCTGTGCCCAGCAGTGCTCCGGGCGCTGCCGTGGCAAGTCCCCCAGTGACTGCTGCCACAACCAGTGTGCTGCAGGCTGCACAGGCCCCCGGGAGAGCGACTGCCTGGTCTGCCGCAAATTCCGAGACGAAGCCACGTGCAAGGACACCTGCCCCCCACTCATGCTCTACAACCCCACCACGTACCAGATGGATGTGAACCCCGAGGGCAAATACAGCTTTGGTGCCACCTGCGTGAAGAAGTGTCCCCGTAATTATGTGGTGACAGATCACGGCTCGTGCGTCCGAGCCTGTGGGGCCGACAGCTATGAGATGGAGGAAGACGGCGTCCGCAAGTGTAAGAAGTGCGAAGGGCCTTGCCGCAAAGTGTGTAACGGAATAGGTATTGGTGAATTTAAAGACTCACTCTCCATAAATGCTACGAATATTAAACACTTCAAAAACTGCACCTCCATCAGTGGCGATCTCCACATCCTGCCGGTGGCATTTAGGGGTGACTCCTTCACACATACTCCTCCTCTGGATCCACAGGAACTGGATATTCTGAAAACCGTAAAGGAAATCACAGGGTTTTTGCTGATTCAGGCTTGGCCTGAAAACAGGACGGACCTCCATGCCTTTGAGAACCTAGAAATCATACGCGGCAGGACCAAGCAACATGGTCAGTTTTCTCTTGCAGTCGTCAGCCTGAACATAACATCCTTGGGATTACGCTCCCTCAAGGAGATAAGTGATGGAGATGTGATAATTTCAGGAAACAAAAATTTGTGCTATGCAAATACAATAAACTGGAAAAAACTGTTTGGGACCTCCGGTCAGAAAACCAAAATTATAAGCAACAGAGGTGAAAACAGCTGCAAGGCCACAGGCCAGGTCTGCCATGCCTTGTGCTCCCCCGAGGGCTGCTGGGGCCCGGAGCCCAGGGACTGCGTCTCTTGCCGGAATGTCAGCCGAGGCAGGGAATGCGTGGACAAGTGCAACCTTCTGGAGGGTGAGCCAAGGGAGTTTGTGGAGAACTCTGAGTGCATACAGTGCCACCCAGAGTGCCTGCCTCAGGCCATGAACATCACCTGCACAGGACGGGGACCAGACAACTGTATCCAGTGTGCCCACTACATTGACGGCCCCCACTGCGTCAAGACCTGCCCGGCAGGAGTCATGGGAGAAAACAACACCCTGGTCTGGAAGTACGCAGACGCCGGCCATGTGTGCCACCTGTGCCATCCAAACTGCACCTACGGATGCACTGGGCCAGGTCTTGAAGGCTGTCCAACGAATGGGCCTAAGATCCCGTCCATCGCCACTGGGATGGTGGGGGCCCTCCTCTTGCTGCTGGTGGTGGCCCTGGGGATCGGCCTCTTCATGCGAAGGCGCCACATCGTTCGGAAGCGCACGCTGCGGAGGCTGCTGCAGGAGAGGGAGCTTGTGGAGCCTCTTACACCCAGTGGAGAAGCTCCCAACCAAGCTCTCTTGAGGATCTTGAAGGAAACTGAATTCAAAAAGATCAAAGTGCTGGGCTCCGGTGCGTTCGGCACGGTGTATAAGGGACTCTGGATCCCAGAAGGTGAGAAAGTTAAAATTCCCGTCGCTATCAAGGAATTAAGAGAAGCAACATCTCCGAAAGCCAACAAGGAAATCCTCGATGAAGCCTACGTGATGGCCAGCGTGGACAACCCCCACGTGTGCCGCCTGCTGGGCATCTGCCTCACCTCCACCGTGCAGCTCATCACGCAGCTCATGCCCTTCGGCTGCCTCCTGGACTATGTCCGGGAACACAAAGACAATATTGGCTCCCAGTACCTGCTCAACTGGTGTGTGCAGATCGCAAAGGGCATGAACTACTTGGAGGACCGTCGCTTGGTGCACCGCGACCTGGCAGCCAGGAACGTACTGGTGAAAACACCGCAGCATGTCAAGATCACAGATTTTGGGCTGGCCAAACTGCTGGGTGCGGAAGAGAAAGAATACCATGCAGAAGGAGGCAAAGTGCCTATCAAGTGGATGGCATTGGAATCAATTTTACACAGAATCTATACCCACCAGAGTGATGTCTGGAGCTACGGGGTGACTGTTTGGGAGTTGATGACCTTTGGATCCAAGCCATATGACGGAATCCCTGCCAGCGAGATCTCCTCCATCCTGGAGAAAGGAGAACGCCTCCCTCAGCCACCCATATGTACCATCGATGTCTACATGATCATGGTCAAGTGCTGGATGATAGACGCAGATAGTCGCCCAAAGTTCCGTGAGTTGATCATCGAATTCTCCAAAATGGCCCGAGACCCCCAGCGCTACCTTGTCATTCAGGGGGATGAAAGAATGCATTTGCCAAGTCCTACAGACTCCAACTTCTACCGTGCCCTGATGGATGAAGAAGACATGGACGACGTGGTGGATGCCGACGAGTACCTCATCCCACAGCAGGGCTTCTTCAGCAGCCCCTCCACGTCACGGACTCCCCTCCTGAGCTCTCTGAGTGCAACCAGCAACAATTCCACCGTGGCTTGCATTGATAGAAATGGGCTGCAAAGCTGTCCCATCAAGGAAGACAGCTTCTTGCAGCGATACAGCTCAGACCCCACAGGCGCCTTGACTGAGGACAGCATAGACGACACCTTCCTCCCAGTGCCTGAATACATAAACCAGTCCGTTCCCAAAAGGCCCGCTGGCTCTGTGCAGAATCCTGTCTATCACAATCAGCCTCTGAACCCCGCGCCCAGCAGAGACCCACACTACCAGGACCCCCACAGCACTGCAGTGGGCAACCCCGAGTATCTCAACACTGTCCAGCCCACCTGTGTCAACAGCACATTCGACAGCCCTGCCCACTGGGCCCAGAAAGGCAGCCACCAAATTAGCCTGGACAACCCTGACTACCAGCAGGACTTCTTTCCCAAGGAAGCCAAGCCAAATGGCATCTTTAAGGGCTCCACAGCTGAAAATGCAGAATACCTAAGGGTCGCGCCACAAAGCAGTGAATTTATTGGAGCATGA\n",
      "3,633 bps.\n",
      "\n",
      "protein:\n",
      "MRPSGTAGAALLALLAALCPASRALEEKKVCQGTSNKLTQLGTFEDHFLSLQRMFNNCEVVLGNLEITYVQRNYDLSFLKTIQEVAGYVLIALNTVERIPLENLQIIRGNMYYENSYALAVLSNYDANKTGLKELPMRNLQEILHGAVRFSNNPALCNVESIQWRDIVSSDFLSNMSMDFQNHLGSCQKCDPSCPNGSCWGAGEENCQKLTKIICAQQCSGRCRGKSPSDCCHNQCAAGCTGPRESDCLVCRKFRDEATCKDTCPPLMLYNPTTYQMDVNPEGKYSFGATCVKKCPRNYVVTDHGSCVRACGADSYEMEEDGVRKCKKCEGPCRKVCNGIGIGEFKDSLSINATNIKHFKNCTSISGDLHILPVAFRGDSFTHTPPLDPQELDILKTVKEITGFLLIQAWPENRTDLHAFENLEIIRGRTKQHGQFSLAVVSLNITSLGLRSLKEISDGDVIISGNKNLCYANTINWKKLFGTSGQKTKIISNRGENSCKATGQVCHALCSPEGCWGPEPRDCVSCRNVSRGRECVDKCNLLEGEPREFVENSECIQCHPECLPQAMNITCTGRGPDNCIQCAHYIDGPHCVKTCPAGVMGENNTLVWKYADAGHVCHLCHPNCTYGCTGPGLEGCPTNGPKIPSIATGMVGALLLLLVVALGIGLFMRRRHIVRKRTLRRLLQERELVEPLTPSGEAPNQALLRILKETEFKKIKVLGSGAFGTVYKGLWIPEGEKVKIPVAIKELREATSPKANKEILDEAYVMASVDNPHVCRLLGICLTSTVQLITQLMPFGCLLDYVREHKDNIGSQYLLNWCVQIAKGMNYLEDRRLVHRDLAARNVLVKTPQHVKITDFGLAKLLGAEEKEYHAEGGKVPIKWMALESILHRIYTHQSDVWSYGVTVWELMTFGSKPYDGIPASEISSILEKGERLPQPPICTIDVYMIMVKCWMIDADSRPKFRELIIEFSKMARDPQRYLVIQGDERMHLPSPTDSNFYRALMDEEDMDDVVDADEYLIPQQGFFSSPSTSRTPLLSSLSATSNNSTVACIDRNGLQSCPIKEDSFLQRYSSDPTGALTEDSIDDTFLPVPEYINQSVPKRPAGSVQNPVYHNQPLNPAPSRDPHYQDPHSTAVGNPEYLNTVQPTCVNSTFDSPAHWAQKGSHQISLDNPDYQQDFFPKEAKPNGIFKGSTAENAEYLRVAPQSSEFIGA\n",
      "1,210 AAs.\n",
      "\n",
      "5'UTR:\n",
      "AGACGTCCGGGCAGCCCCCGGCGCAGCGCGGCCGCAGCAGCCTCCGCCCCCCGCACGGTGTGAGCGCCCGACGCGGCCGAGGCGGCCGGAGTCCCGAGCTAGCCCCGGCGGCCGCCGCCGCCCAGACCGGACGACAGGCCACCTCGTCGGCGTCCGCCCGAGTCCCCGCCTCGCCGCCAACGCCACAACCACCGCGCACGGCCCCCTGACTCCGTCCAGTATTGATCGGGAGAGCCGGAGCGAGCTCTTCGGGGAGCAGCG\n",
      "261 bps.\n",
      "\n",
      "3'UTR=\n",
      "CCACGGAGGATAGTATGAGCCCTAAAAATCCAGACTCTTTCGATACCCAGGACCAAGCCACAGCAGGTCCTCCATCCCAACAGCCATGCCCGCATTAGCTCTTAGACCCACAGACTGGTTTTGCAACGTTTACACCGACTAGCCAGGAAGTACTTCCACCTCGGGCACATTTTGGGAAGTTGCATTCCTTTGTCTTCAAACTGTGAAGCATTTACAGAAACGCATCCAGCAAGAATATTGTCCCTTTGAGCAGAAATTTATCTTTCAAAGAGGTATATTTGAAAAAAAAAAAAAGTATATGTGAGGATTTTTATTGATTGGGGATCTTGGAGTTTTTCATTGTCGCTATTGATTTTTACTTCAATGGGCTCTTCCAACAAGGAAGAAGCTTGCTGGTAGCACTTGCTACCCTGAGTTCATCCAGGCCCAACTGTGAGCAAGGAGCACAAGCCACAAGTCTTCCAGAGGATGCTTGATTCCAGTGGTTCTGCTTCAAGGCTTCCACTGCAAAACACTAAAGATCCAAGAAGGCCTTCATGGCCCCAGCAGGCCGGATCGGTACTGTATCAAGTCATGGCAGGTACAGTAGGATAAGCCACTCTGTCCCTTCCTGGGCAAAGAAGAAACGGAGGGGATGGAATTCTTCCTTAGACTTACTTTTGTAAAAATGTCCCCACGGTACTTACTCCCCACTGATGGACCAGTGGTTTCCAGTCATGAGCGTTAGACTGACTTGTTTGTCTTCCATTCCATTGTTTTGAAACTCAGTATGCTGCCCCTGTCTTGCTGTCATGAAATCAGCAAGAGAGGATGACACATCAAATAATAACTCGGATTCCAGCCCACATTGGATTCATCAGCATTTGGACCAATAGCCCACAGCTGAGAATGTGGAATACCTAAGGATAGCACCGCTTTTGTTCTCGCAAAAACGTATCTCCTAATTTGAGGCTCAGATGAAATGCATCAGGTCCTTTGGGGCATAGATCAGAAGACTACAAAAATGAAGCTGCTCTGAAATCTCCTTTAGCCATCACCCCAACCCCCCAAAATTAGTTTGTGTTACTTATGGAAGATAGTTTTCTCCTTTTACTTCACTTCAAAAGCTTTTTACTCAAAGAGTATATGTTCCCTCCAGGTCAGCTGCCCCCAAACCCCCTCCTTACGCTTTGTCACACAAAAAGTGTCTCTGCCTTGAGTCATCTATTCAAGCACTTACAGCTCTGGCCACAACAGGGCATTTTACAGGTGCGAATGACAGTAGCATTATGAGTAGTGTGGAATTCAGGTAGTAAATATGAAACTAGGGTTTGAAATTGATAATGCTTTCACAACATTTGCAGATGTTTTAGAAGGAAAAAAGTTCCTTCCTAAAATAATTTCTCTACAATTGGAAGATTGGAAGATTCAGCTAGTTAGGAGCCCACCTTTTTTCCTAATCTGTGTGTGCCCTGTAACCTGACTGGTTAACAGCAGTCCTTTGTAAACAGTGTTTTAAACTCTCCTAGTCAATATCCACCCCATCCAATTTATCAAGGAAGAAATGGTTCAGAAAATATTTTCAGCCTACAGTTATGTTCAGTCACACACACATACAAAATGTTCCTTTTGCTTTTAAAGTAATTTTTGACTCCCAGATCAGTCAGAGCCCCTACAGCATTGTTAAGAAAGTATTTGATTTTTGTCTCAATGAAAATAAAACTATATTCATTTCCACTCTATTATGCTCTCAAATACCCCTAAGCATCTATACTAGCCTGGTATGGGTATGAAAGATACAAAGATAAATAAAACATAGTCCCTGATTCTAAGAAATTCACAATTTAGCAAAGGAAATGGACTCATAGATGCTAACCTTAAAACAACGTGACAAATGCCAGACAGGACCCATCAGCCAGGCACTGTGAGAGCACAGAGCAGGGAGGTTGGGTCCTGCCTGAGGAGACCTGGAAGGGAGGCCTCACAGGAGGATGACCAGGTCTCAGTCAGCGGGGAGGTGGAAAGTGCAGGTGCATCAGGGGCACCCTGACCGAGGAAACAGCTGCCAGAGGCCTCCACTGCTAAAGTCCACATAAGGCTGAGGTCAGTCACCCTAAACAACCTGCTCCCTCTAAGCCAGGGGATGAGCTTGGAGCATCCCACAAGTTCCCTAAAAGTTGCAGCCCCCAGGGGGATTTTGAGCTATCATCTCTGCACATGCTTAGTGAGAAGACTACACAACATTTCTAAGAATCTGAGATTTTATATTGTCAGTTAACCACTTTCATTATTCATTCACCTCAGGACATGCAGAAATATTTCAGTCAGAACTGGGAAACAGAAGGACCTACATTCTGCTGTCACTTATGTGTCAAGAAGCAGATGATCGATGAGGCAGGTCAGTTGTAAGTGAGTCACATTGTAGCATTAAATTCTAGTATTTTTGTAGTTTGAAACAGTAACTTAATAAAAGAGCAAAAGCTATTCTAGCTTTCTTCTTCATATTTTAATTTTCCACCATAAAGTTTAGTTGCTAAATTCTATTAATTTTAAGATTGTGCTTCCCAAAATAGTTCTCACTTCATCTGTCCAGGGAGGCACAGTTCTGTCTGGTAGAAGCCGCAAAGCCCTTAGCCTCTTCACGGATCTGGCGACTGTGATGGGCAGGTCAGGAGAGGAGCTGCCCAAAGTCCCATGATTTTCACCTAACAGCCCTGATCAGTCAGTACTCAAAGCTTGGACTCCATCCCTGAAGGTCTTCCTGATTGATAGCCTGGCCTTAATACCCTACAGAAAGCCTGTCCATTGGCTGTTTCTTCCTCAGTCAGTTCCTGGAAGACCTTACCCCATGACCCCAGCTTCAGATGTGGTCTTTGGAAACAGAGGTCGAAGGAAAGTAAGGAGCTGAGAGCTCACATTCATAGGTGCCGCCAGCCTTCGTGCATCTTCTTGCATCATCTCTAAGGAGCTCCTCTAATTACACCATGCCCGTCACCCCATGAGGGATCAGAGAAGGGATGAGTCTTCTAAACTCTATATTCGCTGTGAGTCCAGGTTGTAAGGGGGAGCACTGTGGATGCATCCTATTGCACTCCAGCTGATGACACCAAAGCTTAGGTGTTTGCTGAAAGTTCTTGATGTTGTGACTTACCACCCCTGCCTCACAACTGCAGACATAAGGGGACTATGGATTGCTTAGCAGGAAAGGCACTGGTTCTCAAGGGCGGCTGCCCTTGGGAATCTTCTGGTCCCAACCAGAAAGACTGTGGCTTGATTTTCTCAGGTGCAGCCCAGCCGTAGGGCCTTTTCAGAGCACCCCCTGGTTATTGCAACATTCATCAAAGTTTCTAGAACCTCTGGCCTAAAGGAAGGGCCTGGTGGGATCTACTTGGCACTCGCTGGGGGGCCACCCCCCAGTGCCACTCTCACTAGGCCTCTGATTGCACTTGTGTAGGATGAAGCTGGTGGGTGATGGGAACTCAGCACCTCCCCTCAGGCAGAAAAGAATCATCTGTGGAGCTTCAAAAGAAGGGGCCTGGAGTCTCTGCAGACCAATTCAACCCAAATCTCGGGGGCTCTTTCATGATTCTAATGGGCAACCAGGGTTGAAACCCTTATTTCTAGGGTCTTCAGTTGTACAAGACTGTGGGTCTGTACCAGAGCCCCCGTCAGAGTAGAATAAAAGGCTGGGTAGGGTAGAGATTCCCATGTGCAGTGGAGAGAACAATCTGCAGTCACTGATAAGCCTGAGACTTGGCTCATTTCAAAAGCGTTCAATTCATCCTCACCAGCAGTTCAGCTGGAAAGGGGCAAATACCCCCACCTGAGCTTTGAAAACGCCCTGGGACCCTCTGCATTCTCTAAGTAAGTTATAGAAACCAGTCTCTTCCCTCCTTTGTGAGTGAGCTGCTATTCCACGTAGGCAACACCTGTTGAAATTGCCCTCAATGTCTACTCTGCATTTCTTTCTTGTGATAAGCACACACTTTTATTGCAACATAATGATCTGCTCACATTTCCTTGCCTGGGGGCTGTAAAACCTTACAGAACAGAAATCCTTGCCTCTTTCACCAGCCACACCTGCCATACCAGGGGTACAGCTTTGTACTATTGAAGACACAGACAGGATTTTTAAATGTAAATCTATTTTTGTAACTTTGTTGCGGGATATAGTTCTCTTTATGTAGCACTGAACTTTGTACAATATATTTTTAGAAACTCATTTTTCTACTAAAACAAACACAGTTTACTTTAGAGAGACTGCAATAGAATCAAAATTTGAAACTGAAATCTTTGTTTAAAAGGGTTAAGTTGAGGCAAGAGGAAAGCCCTTTCTCTCTCTTATAAAAAGGCACAACCTCATTGGGGAGCTAAGCTAGGTCATTGTCATGGTGAAGAAGAGAAGCATCGTTTTTATATTTAGGAAATTTTAAAAGATGATGGAAAGCACATTTAGCTTGGTCTGAGGCAGGTTCTGTTGGGGCAGTGTTAATGGAAAGGGCTCACTGTTGTTACTACTAGAAAAATCCAGTTGCATGCCATACTCTCATCATCTGCCAGTGTAACCCTGTACATGTAAGAAAAGCAATAACATAGCACTTTGTTGGTTTATATATATAATGTGACTTCAATGCAAATTTTATTTTTATATTTACAATTGATATGCATTTACCAGTATAAACTAGACATGTCTGGAGAGCCTAATAATGTTCAGCACACTTTGGTTAGTTCACCAACAGTCTTACCAAGCCTGGGCCCAGCCACCCTAGAGAAGTTATTCAGCCCTGGCTGCAGTGACATCACCTGAGGAGCTTTTAAAAGCTTGAAGCCCAGCTACACCTCAGACCGATTAAACGCAAATCTCTGGGGCTGAAACCCAAGCATTCGTAGTTTTTAAAGCTCCTGAGGTCATTCCAATGTGCGGCCAAAGTTGAGAACTACTGGCCTAGGGATTAGCCACAAGGACATGGACTTGGAGGCAAATTCTGCAGGTGTATGTGATTCTCAGGCCTAGAGAGCTAAGACACAAAGACCTCCACATCTGTCGCTGAGAGTCAAGAACCTGAACAGAGTTTCCATGAAGGTTCTCCAAGCACTAGAAGGGAGAGTGTCTAAACAATGGTTGAAAAGCAAAGGAAATATAAAACAGACACCTCTTTCCATTTCCTAAGGTTTCTCTCTTTATTAAGGGTGGACTAGTAATAAAATATAATATTCTTGCTGCTTATGCAGCTGACATTGTTGCCCTCCCTAAAGCAACCAAGTAGCCTTTATTTCCCACAGTGAAAGAAAACGCTGGCCTATCAGTTACATTACAAAAGGCAGATTTCAAGAGGATTGAGTAAGTAGTTGGATGGCTTTCATAAAAACAAGAATTCAAGAAGAGGATTCATGCTTTAAGAAACATTTGTTATACATTCCTCACAAATTATACCTGGGATAAAAACTATGTAGCAGGCAGTGTGTTTTCCTTCCATGTCTCTCTGCACTACCTGCAGTGTGTCCTCTGAGGCTGCAAGTCTGTCCTATCTGAATTCCCAGCAGAAGCACTAAGAAGCTCCACCCTATCACCTAGCAGATAAAACTATGGGGAAAACTTAAATCTGTGCATACATTTCTGGATGCATTTACTTATCTTTAAAAAAAAAGGAATCCTATGACCTGATTTGGCCACAAAAATAATCTTGCTGTACAATACAATCTCTTGGAAATTAAGAGATCCTATGGATTTGATGACTGGTATTAGAGGTGACAATGTAACCGATTAACAACAGACAGCAATAACTTCGTTTTAGAAACATTCAAGCAATAGCTTTATAGCTTCAACATATGGTACGTTTTAACCTTGAAAGTTTTGCAATGATGAAAGCAGTATTTGTACAAATGAAAAGCAGAATTCTCTTTTATATGGTTTATACTGTTGATCAGAAATGTTGATTGTGCATTGAGTATTAAAAAATTAGATGTATATTATTCATTGTTCTTTACTCCTGAGTACCTTATAATAATAATAATGTATTCTTTGTTAACAA\n",
      "6,011 bps.\n",
      "\n",
      "Exon #5:\n",
      "GCCAAAAGTGTGATCCAAGCTGTCCCAATGGGAGCTGCTGGGGTGCAGGAGAGGAGAACTGCCAGAAAC\n",
      "Exon_5:chr7:55151294:55151362\n",
      "\n",
      "Intron #5 contains 1,183 bps.\n",
      "INTRON_5:CHR7:55151363:55152545\n",
      "\n",
      "A modified transcript containing exons 1, 3 and introns 2 contains 1,400 bps.\n"
     ]
    }
   ],
   "source": [
    "pre_mRNA_seq = gA.seq(transcript_id).upper()\n",
    "print(f\"The pre-mRNA sequence contains {len(pre_mRNA_seq):,} bps.\")\n",
    "\n",
    "rna_seq = gA.rna(transcript_id).upper()\n",
    "print(f\"\\nrna:\\n{rna_seq}\\n{len(rna_seq):,} bps.\")\n",
    "\n",
    "orf_seq = gA.ORF(transcript_id).upper()\n",
    "print(f\"\\norf:\\n{orf_seq}\\n{len(orf_seq):,} bps.\")\n",
    "\n",
    "aa_seq = gA.AA(transcript_id)\n",
    "print(f\"\\nprotein:\\n{aa_seq}\\n{len(aa_seq):,} AAs.\")\n",
    "\n",
    "utr5_seq = gA.UTR5(transcript_id).upper()\n",
    "print(f\"\\n5'UTR:\\n{utr5_seq}\\n{len(utr5_seq):,} bps.\")\n",
    "\n",
    "utr3_seq = gA.UTR3(transcript_id).upper()\n",
    "print(f\"\\n3'UTR=\\n{utr3_seq}\\n{len(utr3_seq):,} bps.\")\n",
    "\n",
    "# An Exon sequence\n",
    "exon_number: int = 5\n",
    "exon_seq, seq_info = gA.exon_intron_seq('Exon', exon_number, transcript_id)\n",
    "print(f\"\\nExon #{exon_number}:\\n{exon_seq}\\n{seq_info}\")\n",
    "\n",
    "# An Intron sequence\n",
    "intron_number: int = 5\n",
    "intron_seq, seq_info = gA.exon_intron_seq('Intron', intron_number, transcript_id)\n",
    "print(f\"\\nIntron #{intron_number} contains {len(intron_seq):,} bps.\\n{seq_info.upper()}\")\n",
    "\n",
    "# A modified transcript - generates a sequence based on a list of exons and introns\n",
    "use_exon_list: list[int] = [1, 3]  # exon numbers to use\n",
    "use_intron_list: list[int] = [2]  # intron numbers to use\n",
    "m_seq = gA.modified_transcript(use_exon_list, use_intron_list, transcript_id)\n",
    "print(f\"\\nA modified transcript containing exons {', '.join(map(str,use_exon_list))} and introns {', '.join(map(str,use_intron_list))} contains {len(m_seq.upper()):,} bps.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Queries\n",
    "All positions below are 1-based positions (unless otherwise indicated)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start_codon_info=(55019278, 262)\n",
      "stop_codon_info=(55205615, 3892)\n"
     ]
    }
   ],
   "source": [
    "# query the start and stop codon positions in the chromosome and in the RNA\n",
    "if (start_end_stop := gA.start_and_stop_codons_pos(transcript_id)) is None:\n",
    "    raise ValueError(f\"{transcript_id=} not recognized !!\")\n",
    "start_codon_info, stop_codon_info = start_end_stop\n",
    "# each _info contains (chromosome_position_of_first_bp_of_codon, RNA_position_of_first_bp_of_codon)\n",
    "print(f\"{start_codon_info=}\\n{stop_codon_info=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'start': {'exon_number': 1, 'exon_offset': 261}, 'end': {'exon_number': 28, 'exon_offset': 361}}\n"
     ]
    }
   ],
   "source": [
    "# query the start and end ORF positions in the exons\n",
    "# (i.e., the exon index and the offset within the exon of the \n",
    "# first bp of the start codon and the last bp of the stop codon)\n",
    "orf_start_end_info = gA.orf_start_and_end_exon_info(transcript_id)\n",
    "# here, 'exon_number' value is 1-based and 'exon_offset' is 0-based.\n",
    "print(orf_start_end_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'region': 'Exon_11', 'region_pos': 1, 'dist_from_region_boundary': (0, 90), 'segment': 'ORF', 'pos_in_segment': 1208, 'NT': 'G', 'codon_number': 403, 'nt_in_codon': 2, 'codon': 'GGG', 'aa': 'G'}\n"
     ]
    }
   ],
   "source": [
    "# query a chromosome position\n",
    "chrm_pos: int = 55_157_663\n",
    "# ---------------------------\n",
    "chrm_info = gA.chrm_pos_info(transcript_id, chrm_pos)\n",
    "print(chrm_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrm_pos=55,211,628 --> rna_pos=(9905, 'A')\n"
     ]
    }
   ],
   "source": [
    "# map a chromosome position to the corresponding RNA position\n",
    "chrm_pos: int = 55_211_628 \n",
    "# ------------------------\n",
    "if (rna_pos := gA.chrm_pos2rna_pos(transcript_id, chrm_pos)) is None:\n",
    "    print(f\"{chrm_pos=:,} does not overlap with the RNA of {transcript_id}.\")\n",
    "else:\n",
    "    print(f\"{chrm_pos=:,} --> {rna_pos=:}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rna_pos=685 --> chrm_p=(55143488, 'G')\n"
     ]
    }
   ],
   "source": [
    "# map a RNA position to the chromosome position\n",
    "rna_pos: int = 685\n",
    "# ----------------\n",
    "chrm_p = gA.rna_pos2chrm_pos(transcript_id, rna_pos)\n",
    "# chrm_p is a tuple of the chromosome position and the nucleotide value at that position\n",
    "print(f\"{rna_pos=} --> {chrm_p=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'chrm_pos': 55143488, 'region': 'Exon_3', 'region_pos': 184, 'dist_from_region_boundary': (183, 0), 'segment': 'ORF', 'pos_in_segment': 424, 'NT': 'G', 'codon_number': 142, 'nt_in_codon': 1, 'codon': 'GAA', 'aa': 'E'}\n"
     ]
    }
   ],
   "source": [
    "# query a RNA position\n",
    "rna_pos: int = 685\n",
    "# ----------------\n",
    "chrm_info = gA.rna_pos2chrm_info(transcript_id, rna_pos)\n",
    "print(chrm_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'segment': 'ORF', 'pos_in_segment': 794, 'NT': 'C', 'codon_number': 265, 'nt_in_codon': 2, 'codon': 'CCC', 'aa': 'P'}\n"
     ]
    }
   ],
   "source": [
    "# query an exon position\n",
    "exon_number: int = 7\n",
    "bp_index_in_exon: int = 47\n",
    "# -------------------\n",
    "info = gA.exon_nt_info(transcript_id, exon_number, bp_index_in_exon)\n",
    "print(info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bp #400 in exon #28 corresponds to bp #38 in the 3UTR segment.\n"
     ]
    }
   ],
   "source": [
    "# query the RNA segment (UTR, ORF) of an exon position\n",
    "exon_number: int = 28\n",
    "bp_index_in_exon: int = 400\n",
    "# --------------------------\n",
    "if (exon_segment_info := gA.exon_nt_segment(transcript_id, exon_number, bp_index_in_exon)) is not None:\n",
    "    print(f\"bp #{bp_index_in_exon} in exon #{exon_number} corresponds to bp #{exon_segment_info[1]:,} in the {exon_segment_info[0]} segment.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'codon': 'CAG', 'AA': 'Q', 'codon_exon_pos': ['Exon_4:63', 'Exon_4:64', 'Exon_4:65'], 'codon_chromosome_pos': [55146668, 55146669, 55146670], 'mrna_pos': [748, 749, 750]}\n"
     ]
    }
   ],
   "source": [
    "# query an amino-acid position\n",
    "aa_number: int = 163\n",
    "# ------------------\n",
    "aa_info = gA.aa_exon_info(transcript_id, aa_number)\n",
    "print(aa_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `codon_exon_pos` is a list of 3 positions, each in the form of `<exon number>:<pos>`.\n",
    "`pos` is the 1-based position relative to the start of exon `<exon number`. The corresponding\n",
    "chromosome position is then `exon_start_coordinate+m*(pos-1)`, where `m`=`1` [`m`=-`1`] for genes encoded on\n",
    "the positive [negative] DNA strand (recall that in genes encoded on the negative strand, \n",
    "`exon_start_coordinate`>`exon_end_coordinate`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr7:55152609:G>C --> aa_var='C231S'\n"
     ]
    }
   ],
   "source": [
    "# query an AA variant given a DNA variant defined as chromosome_pos:ref_allele>var_allele\n",
    "ref_allele, var_allele, chromosome_pos = 'G', 'C', 55_152_609\n",
    "# ------------------------------------------------------------\n",
    "if (aa_var := gA.DNA_SNP_mut_to_AA_mut(ref_allele, var_allele, chromosome_pos, transcript_id)) is None:\n",
    "    raise ValueError(\"Verify variant is not DEL or INS and located in the ORF.\")\n",
    "print(f\"chr{gA.chrm}:{chromosome_pos}:{ref_allele}>{var_allele} --> {aa_var=}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aa_var='C231S' corresponds to the following DNA variants:\n",
      "1. codon='TCT', {'start_pos': 55152609, 'reference_allele': 'GC', 'alternative_allele': 'CT'}\n",
      "2. codon='TCC', {'start_pos': 55152609, 'reference_allele': 'G', 'alternative_allele': 'C'}\n",
      "3. codon='TCA', {'start_pos': 55152609, 'reference_allele': 'GC', 'alternative_allele': 'CA'}\n",
      "4. codon='TCG', {'start_pos': 55152609, 'reference_allele': 'GC', 'alternative_allele': 'CG'}\n",
      "5. codon='AGT', {'start_pos': 55152608, 'reference_allele': 'TGC', 'alternative_allele': 'AGT'}\n",
      "6. codon='AGC', {'start_pos': 55152608, 'reference_allele': 'T', 'alternative_allele': 'A'}\n"
     ]
    }
   ],
   "source": [
    "# query (all) DNA variants given an AA variant\n",
    "aa_var: str = 'C231S'\n",
    "# -------------------\n",
    "if (dna_all_muts := gA.AA_mut_to_DNA_SNP_mut(aa_var, transcript_id)) is None:\n",
    "    raise ValueError(\"Verify aa_var is a valid AA variant.\")\n",
    "print(f\"{aa_var=} corresponds to the following DNA variants:\")\n",
    "for i, (codon, var_info) in enumerate(dna_all_muts.items(), start=1):\n",
    "    print(f\"{i}. {codon=}, {var_info}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotating Multiple Genes\n",
    "A faster instantiation of annotation objects, which is relevant when annotating multiple genes, can be done using annotation dataframes that are generated from the annotation file as shown here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First, parse the annotation file into annotation dataframes\n",
    "gff3_dfs = u.ensembl_gff3_df(annotation_full_file)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene=IDH1, Gene name=IDH1, Gene ID=ENSG00000138413\n",
      "Source=Input DataFrame (ID=5434342224, size=(4105687, 9)). (Gene type DataFrame (ID=5434342704, size=(78932, 11)))\n",
      "Positive strand=False.\n",
      "Type=protein_coding, version=14.\n",
      "Description=isocitrate dehydrogenase (NADP(+)) 1 [Source:HGNC Symbol%3BAcc:HGNC:5382]\n",
      "Gene region=chr2:208,236,229-208,266,074.\n",
      "9 transcripts.\n",
      "Gene=BRCA1, Gene name=BRCA1, Gene ID=ENSG00000012048\n",
      "Source=Input DataFrame (ID=5434342224, size=(4105687, 9)). (Gene type DataFrame (ID=5434342704, size=(78932, 11)))\n",
      "Positive strand=False.\n",
      "Type=protein_coding, version=26.\n",
      "Description=BRCA1 DNA repair associated [Source:HGNC Symbol%3BAcc:HGNC:1100]\n",
      "Gene region=chr17:43,044,295-43,170,245.\n",
      "41 transcripts.\n",
      "Gene=EGFR, Gene name=EGFR, Gene ID=ENSG00000146648\n",
      "Source=Input DataFrame (ID=5434342224, size=(4105687, 9)). (Gene type DataFrame (ID=5434342704, size=(78932, 11)))\n",
      "Positive strand=True.\n",
      "Type=protein_coding, version=21.\n",
      "Description=epidermal growth factor receptor [Source:HGNC Symbol%3BAcc:HGNC:3236]\n",
      "Gene region=chr7:55,019,017-55,211,628.\n",
      "13 transcripts.\n"
     ]
    }
   ],
   "source": [
    "# Now instantiate annotation objects (multiple genes) with the annotation dataframes (faster).\n",
    "# Can instantiate in local mode by providing chrm_fasta_file (see above).\n",
    "g_a1 = u.Gene_cls('IDH1', gff3_dfs)\n",
    "g_a1.info()\n",
    "g_a2 = u.Gene_cls('BRCA1', gff3_dfs)\n",
    "g_a2.info()\n",
    "g_a3 = u.Gene_cls('EGFR', gff3_dfs)\n",
    "g_a3.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotating Other Vertebrates Species\n",
    "`geneanot` default species is Homo sapiens (`homo_sapiens`). To annotate other vertebrates species, the variable `species` must be set.\n",
    "\n",
    "## Annotation File\n",
    "The **annotation file signature** is required by the function `update_local_release_to_latest` to download the latest annotation file or check the local release against the latest Ensembl release when annotating other species. If a local annotation file exists (and there is no need to check its release against Ensembl), then this can skipped, and the user must set `annotation_full_file` and `species` when instantiating annotation objects (see below).\n",
    "\n",
    "The **annotation file signature** is identical to the annotation file name, except that the release number is replaced by `XXX`. For example, the signature of the annotation file of Mus musculus is `Mus_musculus.GRCm39.XXX.gff3.gz`. \n",
    "\n",
    "In general, the annotation file name can be found:\n",
    "- by invoking the function `suggested_annotation_file_name(species=species)` as shown below. Or,\n",
    "- in [Ensembl](http://www.ensembl.org/index.html) -> `--Select a species--` (under `All genomes`) -> in the `Gene annotation` section select `GFF3`\n",
    "\n",
    "Following is an example of annotating `Mus musculus` genes and transcripts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotation file name: Mus_musculus.GRCm39.113.gff3.gz.\n",
      "Change the release numnber (113) to 'XXX' to get the signature: Mus_musculus.GRCm39.XXX.gff3.gz\n"
     ]
    }
   ],
   "source": [
    "species: str = 'mus_musculus'\n",
    "\n",
    "# The suggest annotation file name\n",
    "if (anot_info := u.suggested_annotation_file_name(species=species)) is None:\n",
    "    print(f\"Error in retrieving annotation file name. Please check manually for the {species} GFF3 annotation file name in Ensembl.\")\n",
    "else:\n",
    "    suggest_annotation_file_name, assembly_n, release_n = anot_info\n",
    "    suggested_annotation_signature: str = suggest_annotation_file_name.replace(release_n, 'XXX')\n",
    "    print(f\"Annotation file name: {suggest_annotation_file_name}.\\nChange the release numnber ({release_n}) to 'XXX' to get the signature: {suggested_annotation_signature}\")\n",
    "\n",
    "# Annotation file signature\n",
    "annotation_file_signature: str = suggested_annotation_signature\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1 annotation files in ../AnnotationDB. releases=[113], maximal version = 113.\n",
      "ensembl_version=113.\n"
     ]
    }
   ],
   "source": [
    "# Informative - this can be skipped.\n",
    "# Check for local Ensembl annotation files\n",
    "if (local_files := u.get_annotation_files_info_in_folder(Annotation_folder, gff3_pattern=annotation_file_signature)):\n",
    "    releases = list(local_files.keys())\n",
    "    max_ver = max(releases)\n",
    "    print(f\"Found {len(local_files)} annotation files in {Annotation_folder}. {releases=}, maximal version = {max_ver}.\")\n",
    "else:\n",
    "    print(f\"No annotation file in {Annotation_folder}. Need to download from Ensembl.\")\n",
    "\n",
    "# check Ensembl version - Informative\n",
    "ensembl_version = u.get_ensembl_release()\n",
    "print(f\"{ensembl_version=}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Local annotation releases found: 113. Latest is 113 (Mus_musculus.GRCm39.113.gff3.gz).\n",
      "Ensembl latest release is 113.\n",
      "Ensembl and local releases match.\n",
      "Annotation file to use: ../AnnotationDB/Mus_musculus.GRCm39.113.gff3.gz.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# This is required only if want to download annotation file, or check local version against Ensembl latest version.\n",
    "enable_download: bool = False  # enables/disables annotation file download from Ensembl only when local version < ensembl version.\n",
    "                               # If Annotation_folder contains no anotation file, download is done regardless of this setting.\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "download_done, ensembl_file, local_file = u.update_local_release_to_latest(Annotation_folder, \n",
    "                                                                           enable_download=enable_download, \n",
    "                                                                           gff3_pattern=annotation_file_signature,\n",
    "                                                                           species=species)\n",
    "annotation_full_file = Annotation_folder / (ensembl_file if download_done else local_file)\n",
    "print(f\"Annotation file to use: {annotation_full_file}.\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading ../AnnotationDB/Mus_musculus.GRCm39.113.gff3.gz to a dataframe ... Done.\n"
     ]
    }
   ],
   "source": [
    "# Instantiate (in remote access mode)\n",
    "# Can instantiate in local mode by providing chrm_fasta_file (see above)\n",
    "gene: str = 'ENSMUSG00000017167' # set a gene name or a gene ID\n",
    "verbose: bool = True  # set to True to get infromation prints\n",
    "# -----------------------------------------------------------------\n",
    "gAm = u.Gene_cls(gene, annotation_full_file, species=species, verbose=verbose)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene=ENSMUSG00000017167, Gene name=Cntnap1, Gene ID=ENSMUSG00000017167\n",
      "Source=../AnnotationDB/Mus_musculus.GRCm39.113.gff3.gz\n",
      "Positive strand=True.\n",
      "Type=protein_coding, version=7.\n",
      "Description=contactin associated protein-like 1 [Source:MGI Symbol%3BAcc:MGI:1858201]\n",
      "Gene region=chr11:101,061,349-101,081,550.\n",
      "2 transcripts.\n",
      "\n",
      "ENSMUSG00000017167 contains 2 transcripts.\n",
      "ENSMUSG00000017167 transcript IDs:\n",
      " 1. ENSMUST00000138942\n",
      " 2. ENSMUST00000103109\n",
      "\n",
      "Gene start = 101,061,349, gene end = 101,081,550\n"
     ]
    }
   ],
   "source": [
    "gAm.info()\n",
    "\n",
    "print(f\"\\n{gene} contains {len(gAm)} transcripts.\")\n",
    "gAm.show_all_transcript_IDs()\n",
    "\n",
    "print(f\"\\nGene start = {gAm.gene_start:,}, gene end = {gAm.gene_end:,}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "transcript_id='ENSMUST00000103109' annotation:\n",
      "start site (TSS) = 101,066,867, end site (TES) = 101,081,550.\n"
     ]
    }
   ],
   "source": [
    "transcript_id: str = 'ENSMUST00000103109'\n",
    "# --------------------------------------\n",
    "print(f\"\\n{transcript_id=} annotation:\")\n",
    "\n",
    "if (ts_te := gAm.transcript_boundaries(transcript_id)) is None:\n",
    "    raise ValueError(f\"{transcript_id} is not a transcript of {gene} !!\")\n",
    "ts, te = ts_te\n",
    "print(f\"start site (TSS) = {ts:,}, end site (TES) = {te:,}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    exon_number             exon_region  exon_size mRNA_NT_region ORF_NT_region ORF_AA_region  next_exon_frame_alignment             exon_ID\n",
      "0             1  (101066867, 101067153)        287       [1, 287]       [1, 70]       [1, 24]                          2  ENSMUSE00000243064\n",
      "1             2  (101068060, 101068161)        102     [288, 389]     [71, 172]      [24, 58]                          2  ENSMUSE00000112951\n",
      "2             3  (101068249, 101068442)        194     [390, 583]    [173, 366]     [58, 122]                          0  ENSMUSE00000243052\n",
      "3             4  (101068825, 101068972)        148     [584, 731]    [367, 514]    [123, 172]                          2  ENSMUSE00001279109\n",
      "4             5  (101069062, 101069265)        204     [732, 935]    [515, 718]    [172, 240]                          2  ENSMUSE00000112949\n",
      "5             6  (101069641, 101069825)        185    [936, 1120]    [719, 903]    [240, 301]                          0  ENSMUSE00000112946\n",
      "6             7  (101070320, 101070463)        144   [1121, 1264]   [904, 1047]    [302, 349]                          0  ENSMUSE00000661290\n",
      "7             8  (101071136, 101071397)        262   [1265, 1526]  [1048, 1309]    [350, 437]                          2  ENSMUSE00000243015\n",
      "8             9  (101072058, 101072207)        150   [1527, 1676]  [1310, 1459]    [437, 487]                          2  ENSMUSE00000397703\n",
      "9            10  (101072464, 101072635)        172   [1677, 1848]  [1460, 1631]    [487, 544]                          1  ENSMUSE00000112958\n",
      "10           11  (101073029, 101073135)        107   [1849, 1955]  [1632, 1738]    [544, 580]                          2  ENSMUSE00000112953\n",
      "11           12  (101073259, 101073378)        120   [1956, 2075]  [1739, 1858]    [580, 620]                          2  ENSMUSE00000112935\n",
      "12           13  (101073709, 101073912)        204   [2076, 2279]  [1859, 2062]    [620, 688]                          2  ENSMUSE00000112944\n",
      "13           14  (101074040, 101074196)        157   [2280, 2436]  [2063, 2219]    [688, 740]                          1  ENSMUSE00000112942\n",
      "14           15  (101074281, 101074408)        128   [2437, 2564]  [2220, 2347]    [740, 783]                          2  ENSMUSE00000112941\n",
      "15           16  (101074678, 101074863)        186   [2565, 2750]  [2348, 2533]    [783, 845]                          2  ENSMUSE00000112936\n",
      "16           17  (101075357, 101075578)        222   [2751, 2972]  [2534, 2755]    [845, 919]                          2  ENSMUSE00000336522\n",
      "17           18  (101075873, 101076112)        240   [2973, 3212]  [2756, 2995]    [919, 999]                          2  ENSMUSE00000112948\n",
      "18           19  (101077335, 101077688)        354   [3213, 3566]  [2996, 3349]   [999, 1117]                          2  ENSMUSE00000112952\n",
      "19           20  (101077826, 101077953)        128   [3567, 3694]  [3350, 3477]  [1117, 1159]                          0  ENSMUSE00000242931\n",
      "20           21  (101079089, 101079182)         94   [3695, 3788]  [3478, 3571]  [1160, 1191]                          2  ENSMUSE00000112939\n",
      "21           22  (101079335, 101079580)        246   [3789, 4034]  [3572, 3817]  [1191, 1273]                          2  ENSMUSE00000112943\n",
      "22           23  (101079690, 101079737)         48   [4035, 4082]  [3818, 3865]  [1273, 1289]                          2  ENSMUSE00000112956\n",
      "23           24  (101080193, 101081550)       1358   [4083, 5440]  [3866, 4158]  [1289, 1385]                          0  ENSMUSE00000661289\n"
     ]
    }
   ],
   "source": [
    "df = gAm.exon_map(transcript_id)\n",
    "print(df.to_string())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Access mode is remote.\n",
      "\n",
      "protein:\n",
      "MMSLRLFSILLATVVSGAWGWGYYGCNEELVGPLYARSLGASSYYGLFTTARFARLHGISGWSPRIGDPNPWLQIDLMKKHRIRAVATQGAFNSWDWVTRYMLLYGDRVDSWTPFYQKGHNATFFGNVNDSAVVRHDLHYHFTARYIRIVPLAWNPRGKIGLRLGIYGCPYTSSILYFDGDDAISYRFQRGASQSLWDVFAFSFKTEEKDGLLLHTEGSQGDYVTLELQGAHLLLHMSLGSSPIQPRPGHTTVSLGGVLNDLSWHYVRVDRYGRDANFTLDGYAHHFVLNGDFERLNLENEIFIGGLVGAARKNLAYRHNFRGCIENVIYNRINIAEMAVMRHSRITFEGNVAFRCLDPVPHPINFGGPHNFVQVPGFPRRGRLAVSFRFRTWDLTGLLLFSHLGDGLGHVELMLSEGQVNVSIAQTGRKKLQFAAGYRLNDGFWHEVNFVAQENHAVISIDDVEGAEVRVSYPLLIRTGTSYFFGGCPKPASRWGCHSNQTAFHGCMELLKVDGQLVNLTLVEFRKLGYFAEVLFDTCGITDRCSPNMCEHDGRCYQSWDDFICYCELTGYKGVTCHEPLYKESCEAYRLSGKYSGNYTIDPDGSGPLKPFVVYCDIRENRAWTVVRHDRLWTTRVTGSSMDRPFLGAIQYWNASWEEVSALANASQHCEQWIEFSCYNSRLLNTAGGYPYSFWIGRNEEQHFYWGGSQPGIQRCACGLDQSCVDPALHCNCDADQPQWRTDKGLLTFVDHLPVTQVVVGDTNRSNSEAQFFLRPLRCYGDRNSWNTISFHTGAALRFPPIRANHSLDVSFYFRTSAPSGVFLENMGGPFCRWRRPYVRVELNTSRDVVFAFDIGNGDENLTVHSDDFEFNDDEWHLVRAEINVKQARLRVDHRPWVLRPMPLQTYIWLVYDQPLYVGSAELKRRPFVGCLRAMRLNGVTLNLEGRANASEGTFPNCTGHCTHPRFPCFHGGRCVERYSYYTCDCDLTAFDGPYCNHDIGGFFETGTWMRYNLQSALRSAAREFSHMLSRPVPGYEPGYVPGYDTPGYVPGYHGPGYRLPEYPRPGRPVPGYRGPVYNVTGEEVSFSFSTNSAPAVLLYVSSFVRDYMAVLIKEDGTLQLRYQLGTSPYVYQLTTRPVTDGQPHSVNITRVYRNLFIQVDYFPLTEQKFSLLVDSQLDSPKALYLGRVMETGVIDPEIQRYNTPGFSGCLSGVRFNNVAPLKTHFRTPRPMTAELAEAMRVQGELSESNCGAMPRLVSEVPPELDPWYLPPDFPYYHDDGWIAILLGFLVAFLLLGLVGMLVLFYLQNHRYKGSYHTNEPKATHDSHPGGKAPLPPSGPAQAPAPTPAPTQLPTPAPAPAPAPASGPGPRDQNLPQILEESRSE\n",
      "1,385 AAs.\n"
     ]
    }
   ],
   "source": [
    "# protein sequence\n",
    "print(f\"Access mode is {gAm.access_mode()}.\")\n",
    "\n",
    "aa_seq = gAm.AA(transcript_id)\n",
    "print(f\"\\nprotein:\\n{aa_seq}\\n{len(aa_seq):,} AAs.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading ../AnnotationDB/Mus_musculus.GRCm39.113.gff3.gz to a dataframe ... Done.\n",
      "Gene=Drg1, Gene name=Drg1, Gene ID=ENSMUSG00000020457\n",
      "Source=../AnnotationDB/Mus_musculus.GRCm39.113.gff3.gz\n",
      "Positive strand=False.\n",
      "Type=protein_coding, version=14.\n",
      "Description=developmentally regulated GTP binding protein 1 [Source:MGI Symbol%3BAcc:MGI:1343297]\n",
      "Gene region=chr11:3,137,360-3,216,415.\n",
      "7 transcripts.\n",
      "\n",
      "Drg1 contains 7 transcripts.\n",
      "Drg1 transcript IDs:\n",
      " 1. ENSMUST00000159304\n",
      " 2. ENSMUST00000020741\n",
      " 3. ENSMUST00000139430\n",
      " 4. ENSMUST00000142605\n",
      " 5. ENSMUST00000166162\n",
      " 6. ENSMUST00000125424\n",
      " 7. ENSMUST00000132159\n",
      "\n",
      "Gene start = 3,137,360, gene end = 3,216,415\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>region</th>\n",
       "      <th>region_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Exon 1</td>\n",
       "      <td>(3216393, 3216269)</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Intron 1</td>\n",
       "      <td>(3216268, 3214891)</td>\n",
       "      <td>1378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Exon 2</td>\n",
       "      <td>(3214890, 3214767)</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Intron 2</td>\n",
       "      <td>(3214766, 3212665)</td>\n",
       "      <td>2102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Exon 3</td>\n",
       "      <td>(3212664, 3212489)</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Intron 3</td>\n",
       "      <td>(3212488, 3209376)</td>\n",
       "      <td>3113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Exon 4</td>\n",
       "      <td>(3209375, 3209306)</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Intron 4</td>\n",
       "      <td>(3209305, 3206713)</td>\n",
       "      <td>2593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Exon 5</td>\n",
       "      <td>(3206712, 3206543)</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Intron 5</td>\n",
       "      <td>(3206542, 3204643)</td>\n",
       "      <td>1900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Exon 6</td>\n",
       "      <td>(3204642, 3204512)</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Intron 6</td>\n",
       "      <td>(3204511, 3202812)</td>\n",
       "      <td>1700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Exon 7</td>\n",
       "      <td>(3202811, 3202644)</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Intron 7</td>\n",
       "      <td>(3202643, 3202283)</td>\n",
       "      <td>361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Exon 8</td>\n",
       "      <td>(3202282, 3202160)</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Intron 8</td>\n",
       "      <td>(3202159, 3200366)</td>\n",
       "      <td>1794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Exon 9</td>\n",
       "      <td>(3200365, 3199907)</td>\n",
       "      <td>459</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        name              region  region_size\n",
       "0     Exon 1  (3216393, 3216269)          125\n",
       "1   Intron 1  (3216268, 3214891)         1378\n",
       "2     Exon 2  (3214890, 3214767)          124\n",
       "3   Intron 2  (3214766, 3212665)         2102\n",
       "4     Exon 3  (3212664, 3212489)          176\n",
       "5   Intron 3  (3212488, 3209376)         3113\n",
       "6     Exon 4  (3209375, 3209306)           70\n",
       "7   Intron 4  (3209305, 3206713)         2593\n",
       "8     Exon 5  (3206712, 3206543)          170\n",
       "9   Intron 5  (3206542, 3204643)         1900\n",
       "10    Exon 6  (3204642, 3204512)          131\n",
       "11  Intron 6  (3204511, 3202812)         1700\n",
       "12    Exon 7  (3202811, 3202644)          168\n",
       "13  Intron 7  (3202643, 3202283)          361\n",
       "14    Exon 8  (3202282, 3202160)          123\n",
       "15  Intron 8  (3202159, 3200366)         1794\n",
       "16    Exon 9  (3200365, 3199907)          459"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "mRNA:\n",
      "AGTGGTGGCCAGAGGCGCGGCGGGCGCGAGAAGGGGGACCTAACCAGTGTGGAGGCTGGTGTGGCCTTAAGCCCTCCGCCACCATGAGCGGGACCTTAGCGAAGATCGCGGAGATCGAAGCCGAGATGGCTCGGACTCAGAAGAACAAAGCCACAGCTCATCACCTTGGCCTGCTTAAGGCTCGCCTTGCTAAACTTCGAAGAGAACTCATCACTCCTAAAGGTGGCGGCGGCGGGGGGCCAGGAGAAGGTTTTGATGTGGCCAAGACAGGTGATGCTCGAATTGGGTTTGTGGGTTTTCCATCTGTGGGGAAATCAACTCTGCTAAGTAACTTAGCAGGAGTGTATTCTGAAGTGGCAGCCTATGAATTCACTACATTGACCACAGTGCCTGGTGTCATCAGATATAAAGGTGCCAAGATCCAGCTCCTGGATCTCCCGGGTATTATTGAAGGTGCCAAGGATGGGAAAGGTAGAGGTCGTCAAGTCATTGCAGTGGCCCGAACGTGTAATTTGATTTTGATTGTTCTGGATGTCCTAAAACCCTTGGGACATAAGAAGATAATTGAAAATGAGTTGGAAGGCTTTGGGATTCGCTTGAACAGCAAACCCCCCAACATTGGCTTTAAGAAGAAAGACAAGGGAGGCATTAACCTCACAGCCACTTGTCCTCAGAGTGAACTGGATGCTGAAACAGTGAAGAGCATTCTGGCTGAATATAAGATTCATAATGCTGATGTGACTTTGCGTAGTGATGCCACAGCTGACGACCTCATTGACGTGGTCGAAGGGAACAGAGTTTATATTCCCTGTATCTATGTATTAAATAAGATTGATCAGATCTCCATTGAGGAGTTGGATATTATCTATAAGGTGCCTCACTGTGTACCCATCTCTGCTCATCACCGCTGGAATTTTGATGACCTTTTGGAGAAAATCTGGGACTATCTGAAACTAGTGAGAATTTACACCAAACCGAAAGGCCAGTTACCAGATTACACATCTCCAGTAGTGCTGCCTTATTCTAGGACCACTGTGGAAGATTTCTGCATGAAGATTCACAAAAATCTCATCAAAGAATTTAAATATGCTTTGGTCTGGGGTCTGTCTGTGAAACATAATCCTCAGAAAGTGGGTAAAGACCACACATTGGAAGATGAAGACGTCATTCAGATTGTGAAGAAGTGAAGTCCTCTTGCCATCTGCTGGCTGGCCACAGCAGTGTTCCTTATGACTAAGTACCTTCCTCAGCCCCTCTGTCTTTGGCAGCCTCTGGATTGGGATACAGGGATTGATATGGAGACTTATCAAACTGAAACACCACAATTTTTTTCTTTCCGGAGATAGGGTGTTATGTATTTCAGGCTGGCTTTGAACTCTTGCAGCTGAGGCTGGCCTAGAATCCCTTTCATCAGCCTCCATCTCCTCAGTGCTGGGCTTACCACTGTGCACCTTTAGCTCCTCAGCAACACCACTTTCCCTACTTTGCTGTCACTTCTATTGAACTGTATAAAGAAGCTGGTGGGCCCGCTGGCCATGTGTCGTTA\n",
      "1,546 bps.\n",
      "\n",
      "protein:\n",
      "MSGTLAKIAEIEAEMARTQKNKATAHHLGLLKARLAKLRRELITPKGGGGGGPGEGFDVAKTGDARIGFVGFPSVGKSTLLSNLAGVYSEVAAYEFTTLTTVPGVIRYKGAKIQLLDLPGIIEGAKDGKGRGRQVIAVARTCNLILIVLDVLKPLGHKKIIENELEGFGIRLNSKPPNIGFKKKDKGGINLTATCPQSELDAETVKSILAEYKIHNADVTLRSDATADDLIDVVEGNRVYIPCIYVLNKIDQISIEELDIIYKVPHCVPISAHHRWNFDDLLEKIWDYLKLVRIYTKPKGQLPDYTSPVVLPYSRTTVEDFCMKIHKNLIKEFKYALVWGLSVKHNPQKVGKDHTLEDEDVIQIVKK\n",
      "367 AAs.\n"
     ]
    }
   ],
   "source": [
    "# Another example (gene encoded on the negative strand)\n",
    "gene: str = 'Drg1'\n",
    "gAm = u.Gene_cls(gene, annotation_full_file, species=species, verbose=verbose)\n",
    "\n",
    "gAm.info()\n",
    "\n",
    "print(f\"\\n{gene} contains {len(gAm)} transcripts.\")\n",
    "gAm.show_all_transcript_IDs()\n",
    "\n",
    "print(f\"\\nGene start = {gAm.gene_start:,}, gene end = {gAm.gene_end:,}\")\n",
    "\n",
    "transcript_id: str = 'ENSMUST00000020741'\n",
    "display(gAm.exon_intron_map(transcript_id))\n",
    "\n",
    "mrna_seq = gAm.rna(transcript_id)\n",
    "print(f\"\\nmRNA:\\n{mrna_seq}\\n{len(mrna_seq):,} bps.\")\n",
    "\n",
    "aa_seq = gAm.AA(transcript_id)\n",
    "print(f\"\\nprotein:\\n{aa_seq}\\n{len(aa_seq):,} AAs.\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
